{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c01fa68b-1e71-4a5d-a68a-77b4d6e810f7",
   "metadata": {},
   "source": [
    "# Pydantic validation framework for the EMO BON observatory and other metadata log sheets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71bb619d-5b92-4bee-b1a5-b3f0cb364f70",
   "metadata": {},
   "source": [
    "- **pydantic** Data validation using Python type hints.\n",
    "    - [pypi](https://pypi.org/project/pydantic/)\n",
    "    - [Documentation](https://docs.pydantic.dev/latest/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b5bca054-8458-4137-ae94-b960bbe0ad45",
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime\n",
    "import os\n",
    "import sys\n",
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# Add project directory to path\n",
    "PROJECT_DIR = Path.cwd().parents[0]\n",
    "sys.path.append(str(PROJECT_DIR))\n",
    "sys.path.append(str(PROJECT_DIR / \"src\"))\n",
    "\n",
    "# Weird stuff from JupyterHub after I moved modules and notebooks around:\n",
    "# For some reasong CWD is /src/scratch even though this notebook is in /srv/scratch/emo-bon-validation\n",
    "# The terminal also show us to be in /srv/scratch/emo-bon-validation\n",
    "# So...\n",
    "# if os.getcwd() == \"/srv/scratch\":\n",
    "#    os.chdir(\"./emo-bon-data-validation\")\n",
    "# print(f\"CWD is {os.getcwd()}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e601478-27ea-404a-8371-1405051ccf22",
   "metadata": {},
   "source": [
    "## Governance data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5e34558-fb7c-4000-8e8b-24ad71df7eab",
   "metadata": {},
   "source": [
    "#### Read each of the governance CSV files into a Pandas dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "154ba717-c9a9-43a0-b34e-8b7461952135",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 18 entries, 0 to 17\n",
      "Data columns (total 9 columns):\n",
      " #   Column                               Non-Null Count  Dtype \n",
      "---  ------                               --------------  ----- \n",
      " 0   EMBRC Node                           18 non-null     object\n",
      " 1   EMBRC Site                           18 non-null     object\n",
      " 2   EMOBON_observatory_id                18 non-null     object\n",
      " 3   Water Column                         17 non-null     object\n",
      " 4   Soft sediment                        7 non-null      object\n",
      " 5   data_quality_control_threshold_date  18 non-null     object\n",
      " 6   data_quality_control_assignee        18 non-null     object\n",
      " 7   rocrate_profile_uri                  18 non-null     object\n",
      " 8   autogenerate                         18 non-null     int64 \n",
      "dtypes: int64(1), object(8)\n",
      "memory usage: 1.4+ KB\n",
      "This is info() for None\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 19 entries, 0 to 18\n",
      "Data columns (total 22 columns):\n",
      " #   Column                            Non-Null Count  Dtype  \n",
      "---  ------                            --------------  -----  \n",
      " 0   country_code                      19 non-null     object \n",
      " 1   country                           19 non-null     object \n",
      " 2   EMOBON_observatory_name           19 non-null     object \n",
      " 3   EMOBON_observatory_id             19 non-null     object \n",
      " 4   startdate                         19 non-null     object \n",
      " 5   enddate                           2 non-null      object \n",
      " 6   Water_Column                      19 non-null     object \n",
      " 7   Soft_Substrates                   19 non-null     object \n",
      " 8   Hard_Substrates                   19 non-null     object \n",
      " 9   water_site_latitude               18 non-null     float64\n",
      " 10  water_site_longitude              18 non-null     float64\n",
      " 11  sediment_site_latitude            9 non-null      float64\n",
      " 12  sediment_site_longtitude          9 non-null      float64\n",
      " 13  hard_substrates_site1_longitude   9 non-null      object \n",
      " 14  hard_substrates_site1_latitude    9 non-null      object \n",
      " 15  hard_substrates_site2_longtitude  3 non-null      float64\n",
      " 16  hard_substrates_site2_latitude    3 non-null      float64\n",
      " 17  contact_person                    19 non-null     object \n",
      " 18  contact_person_email              19 non-null     object \n",
      " 19  ENA_accession_number_umbrella     19 non-null     object \n",
      " 20  ENA_accession_number_project      19 non-null     object \n",
      " 21  EMOBON_core                       19 non-null     object \n",
      "dtypes: float64(6), object(16)\n",
      "memory usage: 3.4+ KB\n",
      "This is info() for None\n"
     ]
    }
   ],
   "source": [
    "github_path = \"https://raw.githubusercontent.com/emo-bon/governance-data/main/\"\n",
    "file_names = [\n",
    "    \"logsheets.csv\",  # contain the URLs of the googlesheets that are the logsheets\n",
    "    \"observatories.csv\",  # contain information about each observatory\n",
    "    # \"organisations.csv\",         # contain information about the organisations in EMO BON\n",
    "    # \"planned_events.csv\"         # contains information about planned EMO BON events (this file is only used by humans, not by any actions) - DONT CARE\n",
    "    # \"ro-crate-metadata.json\"     # IGNORE\n",
    "]\n",
    "\n",
    "dfs = {}\n",
    "for f in file_names:\n",
    "    df = pd.read_csv(os.path.join(github_path, f))\n",
    "    print(f\"This is info() for {df.info()}\")\n",
    "    dfs[f] = df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2484615f-cab2-48a2-8ca7-999bab0051e3",
   "metadata": {},
   "source": [
    "#### Validate Governance tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b8dc50e0-2d84-4508-8d82-67b90397c39c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from validation_classes import logsheetsModel, observatoriesModel\n",
    "\n",
    "validator_class_paths = {\n",
    "    \"logsheets.csv\": logsheetsModel,\n",
    "    \"observatories.csv\": observatoriesModel,\n",
    "}\n",
    "validation_classes_path = \"./validation_classes\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa819fa6-c492-481f-bc81-218057699342",
   "metadata": {},
   "source": [
    "##### Observatories table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4af3dfa6-8d4c-4749-b9d1-bc00834e1a3b",
   "metadata": {},
   "source": [
    "The observatories validator mostly changes the column names to make them consistent (and spelled correctly), removes blank strings (\"   \") from cells, and reformats the dates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9aa2840b-9b25-4155-bd43-34cea9c81de9",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name = \"observatories.csv\"\n",
    "data = dfs[file_name]  # dfs is dict of pandas df's\n",
    "validator = validator_class_paths[file_name]\n",
    "data_records = data.to_dict(orient=\"records\")\n",
    "validated_rows = [validator(**row).model_dump() for row in data_records]\n",
    "\n",
    "ndf = pd.DataFrame.from_records(validated_rows, index=\"observatory_id\")\n",
    "ndf.to_csv(\n",
    "    PROJECT_DIR / \"validated-data\" / \"governance\" / \"observatories_validated.csv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5ef5ade-e3f4-4742-b1ba-41233e93e80c",
   "metadata": {},
   "source": [
    "##### Logsheets table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7b9e1b9e-8a05-4bf2-aceb-55c5c740e77d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "cdir = os.getcwd()\n",
    "file_name = \"logsheets.csv\"\n",
    "data = dfs[file_name]  # dfs is dict of pandas df's\n",
    "validator = validator_class_paths[file_name]\n",
    "data_records = data.to_dict(orient=\"records\")\n",
    "validated_rows = [validator(**row).model_dump() for row in data_records]\n",
    "\n",
    "ndf = pd.DataFrame.from_records(validated_rows, index=\"observatory_id\")\n",
    "ndf.to_csv(PROJECT_DIR / \"validated-data\" / \"governance\" / \"logsheets_validated.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67fc138c-c96f-4349-843c-ad02276b5f43",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "## Validate logsheets from \"water column\" and \"soft sediments\" for the \"sampling\" and \"measured\" tables\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89b0343f-e36b-4685-ac17-e90a9e53f12d",
   "metadata": {},
   "source": [
    "##### Pulls from the raw Google Sheets does a \"lax\" validation where we correct/coerce everything to a consistent type"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71b41f0d-7d88-46e4-9789-442036904749",
   "metadata": {},
   "source": [
    "### !!! A note about Pandas and integer fields with missing values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b45ba77-48d0-422c-940c-a0c1b396f16a",
   "metadata": {},
   "source": [
    "Pandas will read a raw CSV file and try to determine the type while doing so. If it finds an integer column with missing values they will be NaN's, which of course are floats. Consequently, the default action here is to read the column as a floats by coercing the integer values to float to match the NaNs - this is rarely what you want.\n",
    "\n",
    "However, Pandas does have a [nullable integer type](https://pandas.pydata.org/docs/user_guide/integer_na.html) - `pandas.Int64Dtype()` or it's string alias `\"Int64\"`. You can force Pandas to use this type when reading the CSV file using `pandas.read_csv('file.csv', dtype={\"<int field with missing values>\": \"Int64\"})` the NaN's will be changed to [pandas.NA types](https://pandas.pydata.org/docs/reference/api/pandas.NA.html#pandas.NA).\n",
    "\n",
    "For clarity: this is NOT what happens below: we let the validators deal with it, which coerce the now floats back into ints."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c95e6614-c9ff-470d-8363-8f37a08d34ac",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "Observatory_id ESC68N sheet_link https://docs.google.com/spreadsheets/d/11_Eu0W1-sDiuzKx1cIl6YuxjRHmWezN6u9v3Ly8JZ3A/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='ESC68N' - sampling_strategy='water_column' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/11_Eu0W1-sDiuzKx1cIl6YuxjRHmWezN6u9v3Ly8JZ3A/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/ESC68N_water_column_sampling_validated.csv\n",
      "Observatory_id Bergen sheet_link https://docs.google.com/spreadsheets/d/1HuXHiUJICZrmCrJ4EZDyU5aSCMzDAc1cy_tne5YVPTg/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='Bergen' - sampling_strategy='water_column' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1HuXHiUJICZrmCrJ4EZDyU5aSCMzDAc1cy_tne5YVPTg/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/Bergen_water_column_sampling_validated.csv\n",
      "Observatory_id MBAL4 sheet_link https://docs.google.com/spreadsheets/d/1xfrqraPa0auQ1O-C9RUo68RhxrPCDWkVMCAUbj79AZI/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='MBAL4' - sampling_strategy='water_column' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1xfrqraPa0auQ1O-C9RUo68RhxrPCDWkVMCAUbj79AZI/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/MBAL4_water_column_sampling_validated.csv\n",
      "Observatory_id BPNS sheet_link https://docs.google.com/spreadsheets/d/1mEi4Bd2YR63WD0j54FQ6QkzcUw_As9Wilue9kaXO2DE/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='BPNS' - sampling_strategy='water_column' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1mEi4Bd2YR63WD0j54FQ6QkzcUw_As9Wilue9kaXO2DE/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/BPNS_water_column_sampling_validated.csv\n",
      "Observatory_id ROSKOGO sheet_link https://docs.google.com/spreadsheets/d/1BCu2pbuIS4f-yrw5Kw4uzfO9v6ryZJBtSF5WBLWG_-E/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='ROSKOGO' - sampling_strategy='water_column' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1BCu2pbuIS4f-yrw5Kw4uzfO9v6ryZJBtSF5WBLWG_-E/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/ROSKOGO_water_column_sampling_validated.csv\n",
      "Observatory_id VB sheet_link https://docs.google.com/spreadsheets/d/1kSlTcNfXaCQv7fIKVbBnt5PmRW3_H7Trryh4FJHyLkE/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='VB' - sampling_strategy='water_column' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1kSlTcNfXaCQv7fIKVbBnt5PmRW3_H7Trryh4FJHyLkE/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Discarded 2 records leaving 824.\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/VB_water_column_sampling_validated.csv\n",
      "Observatory_id OOB sheet_link nan\n",
      "Observatory OOB lacks valid sheet URL for water_column\n",
      "Observatory_id EMT21 sheet_link https://docs.google.com/spreadsheets/d/1RutVpduwL1mEed5jC2zLbs0DgkzAwZdsJ27fg3kb0Yg/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='EMT21' - sampling_strategy='water_column' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1RutVpduwL1mEed5jC2zLbs0DgkzAwZdsJ27fg3kb0Yg/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/EMT21_water_column_sampling_validated.csv\n",
      "Observatory_id PiEGetxo sheet_link https://docs.google.com/spreadsheets/d/1KGcishLP5eAn6ZCnixn5wZIez1Y9MgmYoch2s7aefKM/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='PiEGetxo' - sampling_strategy='water_column' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1KGcishLP5eAn6ZCnixn5wZIez1Y9MgmYoch2s7aefKM/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/PiEGetxo_water_column_sampling_validated.csv\n",
      "Observatory_id Plenzia sheet_link https://docs.google.com/spreadsheets/d/13XeETquTvY6F_8Re25E0BAmSxLi0mThDUXYay3Eb5-o/edit?usp=sharing\n",
      "Observatory_id RFormosa sheet_link https://docs.google.com/spreadsheets/d/15f_yqrhvOr1MW-SHuv0JiKf0HRDy81uGZapJANxL54k/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='RFormosa' - sampling_strategy='water_column' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/15f_yqrhvOr1MW-SHuv0JiKf0HRDy81uGZapJANxL54k/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Discarded 10 records leaving 180.\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/RFormosa_water_column_sampling_validated.csv\n",
      "Observatory_id OSD74 sheet_link https://docs.google.com/spreadsheets/d/1VATpzmhwmk8sYkhEjCbDUZHRiYuNGpSRefhXthZRbfQ/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='OSD74' - sampling_strategy='water_column' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1VATpzmhwmk8sYkhEjCbDUZHRiYuNGpSRefhXthZRbfQ/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/OSD74_water_column_sampling_validated.csv\n",
      "Observatory_id AAOT sheet_link https://docs.google.com/spreadsheets/d/1hvLkBwiKTGTJDx19m_8e7qJ2lm9bwLLeVztMpxTLqnk/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='AAOT' - sampling_strategy='water_column' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1hvLkBwiKTGTJDx19m_8e7qJ2lm9bwLLeVztMpxTLqnk/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/AAOT_water_column_sampling_validated.csv\n",
      "Observatory_id NRMCB sheet_link https://docs.google.com/spreadsheets/d/1F4AWv_seI-DQJ_Gp_N2ziwvGyjnc4KC92GGNXrJRek4/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='NRMCB' - sampling_strategy='water_column' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1F4AWv_seI-DQJ_Gp_N2ziwvGyjnc4KC92GGNXrJRek4/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/NRMCB_water_column_sampling_validated.csv\n",
      "Observatory_id HCMR-1 sheet_link https://docs.google.com/spreadsheets/d/13DcVK2mzSxMJoFydSBaIMmj7Td1_JapEvcY2bmZTyLc/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='HCMR-1' - sampling_strategy='water_column' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/13DcVK2mzSxMJoFydSBaIMmj7Td1_JapEvcY2bmZTyLc/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/HCMR-1_water_column_sampling_validated.csv\n",
      "Observatory_id IUIEilat sheet_link https://docs.google.com/spreadsheets/d/1qIwi1nZu4OBiriCzn1fEG1fyG8-3wSVvwzsMiH6JTwk/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='IUIEilat' - sampling_strategy='water_column' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1qIwi1nZu4OBiriCzn1fEG1fyG8-3wSVvwzsMiH6JTwk/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/IUIEilat_water_column_sampling_validated.csv\n",
      "Observatory_id UMF sheet_link https://docs.google.com/spreadsheets/d/1-1VsUUbRtKselxu-y2BghRIrJdaf74SbmvK42ntvd20/edit#gid=2052143695\n",
      "\n",
      "\n",
      "Processing observatory_id='UMF' - sampling_strategy='water_column' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1-1VsUUbRtKselxu-y2BghRIrJdaf74SbmvK42ntvd20/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/UMF_water_column_sampling_validated.csv\n",
      "Observatory_id LMO sheet_link https://docs.google.com/spreadsheets/d/1AvQMYcS0tdNMw6Er8zUarQg1a_wrshhnkTS6RuI1FJQ/edit#gid=2052143695\n",
      "\n",
      "\n",
      "Processing observatory_id='LMO' - sampling_strategy='water_column' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1AvQMYcS0tdNMw6Er8zUarQg1a_wrshhnkTS6RuI1FJQ/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/LMO_water_column_sampling_validated.csv\n",
      "Observatory_id ESC68N sheet_link nan\n",
      "Observatory ESC68N lacks valid sheet URL for soft_sediment\n",
      "Observatory_id Bergen sheet_link nan\n",
      "Observatory Bergen lacks valid sheet URL for soft_sediment\n",
      "Observatory_id MBAL4 sheet_link nan\n",
      "Observatory MBAL4 lacks valid sheet URL for soft_sediment\n",
      "Observatory_id BPNS sheet_link https://docs.google.com/spreadsheets/d/1zc0bZdpl-Eoi35lI_5BGkElbscplyQRyNPLkSgeEyEQ/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='BPNS' - sampling_strategy='soft_sediment' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1zc0bZdpl-Eoi35lI_5BGkElbscplyQRyNPLkSgeEyEQ/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Discarded 11 records leaving 182.\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/BPNS_soft_sediment_sampling_validated.csv\n",
      "Observatory_id ROSKOGO sheet_link https://docs.google.com/spreadsheets/d/1M0ytBWbUDP0YthGtTVtdIaX2cQUxF_uNIJBHIXREoX8/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='ROSKOGO' - sampling_strategy='soft_sediment' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1M0ytBWbUDP0YthGtTVtdIaX2cQUxF_uNIJBHIXREoX8/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/ROSKOGO_soft_sediment_sampling_validated.csv\n",
      "Observatory_id VB sheet_link nan\n",
      "Observatory VB lacks valid sheet URL for soft_sediment\n",
      "Observatory_id OOB sheet_link https://docs.google.com/spreadsheets/d/1J469_ljfxM9NhOEyRmvWQNOXX7dujxBy1opU6ROmDv8/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='OOB' - sampling_strategy='soft_sediment' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1J469_ljfxM9NhOEyRmvWQNOXX7dujxBy1opU6ROmDv8/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/OOB_soft_sediment_sampling_validated.csv\n",
      "Observatory_id EMT21 sheet_link https://docs.google.com/spreadsheets/d/1sBB0x6h-prnUHMcOfms_7qSqTn6zeGVriitTysceoT0/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='EMT21' - sampling_strategy='soft_sediment' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1sBB0x6h-prnUHMcOfms_7qSqTn6zeGVriitTysceoT0/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/EMT21_soft_sediment_sampling_validated.csv\n",
      "Observatory_id PiEGetxo sheet_link nan\n",
      "Observatory PiEGetxo lacks valid sheet URL for soft_sediment\n",
      "Observatory_id Plenzia sheet_link nan\n",
      "Observatory Plenzia lacks valid sheet URL for soft_sediment\n",
      "Observatory_id RFormosa sheet_link https://docs.google.com/spreadsheets/d/17rqrZ-qrDP77SpoEyBFO17wewp98a-tLBWTlBAn6ZCc/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='RFormosa' - sampling_strategy='soft_sediment' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/17rqrZ-qrDP77SpoEyBFO17wewp98a-tLBWTlBAn6ZCc/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Discarded 5 records leaving 90.\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/RFormosa_soft_sediment_sampling_validated.csv\n",
      "Observatory_id OSD74 sheet_link nan\n",
      "Observatory OSD74 lacks valid sheet URL for soft_sediment\n",
      "Observatory_id AAOT sheet_link nan\n",
      "Observatory AAOT lacks valid sheet URL for soft_sediment\n",
      "Observatory_id NRMCB sheet_link https://docs.google.com/spreadsheets/d/1a19q3w2sP7dae8HEzT5TdmfqhRQQd3GItFqiYxWJlxs/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='NRMCB' - sampling_strategy='soft_sediment' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1a19q3w2sP7dae8HEzT5TdmfqhRQQd3GItFqiYxWJlxs/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/NRMCB_soft_sediment_sampling_validated.csv\n",
      "Observatory_id HCMR-1 sheet_link nan\n",
      "Observatory HCMR-1 lacks valid sheet URL for soft_sediment\n",
      "Observatory_id IUIEilat sheet_link nan\n",
      "Observatory IUIEilat lacks valid sheet URL for soft_sediment\n",
      "Observatory_id UMF sheet_link https://docs.google.com/spreadsheets/d/1DpQqo2i4q7RAM0sdm_z2pISa8SDBa9aYhWmUflFuUQs/edit?usp=drive_web&ouid=110054986816469121247\n",
      "\n",
      "\n",
      "Processing observatory_id='UMF' - sampling_strategy='soft_sediment' - sheet_type='sampling'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1DpQqo2i4q7RAM0sdm_z2pISa8SDBa9aYhWmUflFuUQs/gviz/tq?tqx=out:csv&sheet=sampling\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/UMF_soft_sediment_sampling_validated.csv\n",
      "Observatory_id LMO sheet_link nan\n",
      "Observatory LMO lacks valid sheet URL for soft_sediment\n",
      "Observatory_id ESC68N sheet_link https://docs.google.com/spreadsheets/d/11_Eu0W1-sDiuzKx1cIl6YuxjRHmWezN6u9v3Ly8JZ3A/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='ESC68N' - sampling_strategy='water_column' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/11_Eu0W1-sDiuzKx1cIl6YuxjRHmWezN6u9v3Ly8JZ3A/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/ESC68N_water_column_measured_validated.csv\n",
      "Observatory_id Bergen sheet_link https://docs.google.com/spreadsheets/d/1HuXHiUJICZrmCrJ4EZDyU5aSCMzDAc1cy_tne5YVPTg/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='Bergen' - sampling_strategy='water_column' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1HuXHiUJICZrmCrJ4EZDyU5aSCMzDAc1cy_tne5YVPTg/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/Bergen_water_column_measured_validated.csv\n",
      "Observatory_id MBAL4 sheet_link https://docs.google.com/spreadsheets/d/1xfrqraPa0auQ1O-C9RUo68RhxrPCDWkVMCAUbj79AZI/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='MBAL4' - sampling_strategy='water_column' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1xfrqraPa0auQ1O-C9RUo68RhxrPCDWkVMCAUbj79AZI/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/MBAL4_water_column_measured_validated.csv\n",
      "Observatory_id BPNS sheet_link https://docs.google.com/spreadsheets/d/1mEi4Bd2YR63WD0j54FQ6QkzcUw_As9Wilue9kaXO2DE/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='BPNS' - sampling_strategy='water_column' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1mEi4Bd2YR63WD0j54FQ6QkzcUw_As9Wilue9kaXO2DE/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Discarded 15 records leaving 320.\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/BPNS_water_column_measured_validated.csv\n",
      "Observatory_id ROSKOGO sheet_link https://docs.google.com/spreadsheets/d/1BCu2pbuIS4f-yrw5Kw4uzfO9v6ryZJBtSF5WBLWG_-E/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='ROSKOGO' - sampling_strategy='water_column' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1BCu2pbuIS4f-yrw5Kw4uzfO9v6ryZJBtSF5WBLWG_-E/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/ROSKOGO_water_column_measured_validated.csv\n",
      "Observatory_id VB sheet_link https://docs.google.com/spreadsheets/d/1kSlTcNfXaCQv7fIKVbBnt5PmRW3_H7Trryh4FJHyLkE/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='VB' - sampling_strategy='water_column' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1kSlTcNfXaCQv7fIKVbBnt5PmRW3_H7Trryh4FJHyLkE/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/VB_water_column_measured_validated.csv\n",
      "Observatory_id OOB sheet_link nan\n",
      "Observatory OOB lacks valid sheet URL for water_column\n",
      "Observatory_id EMT21 sheet_link https://docs.google.com/spreadsheets/d/1RutVpduwL1mEed5jC2zLbs0DgkzAwZdsJ27fg3kb0Yg/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='EMT21' - sampling_strategy='water_column' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1RutVpduwL1mEed5jC2zLbs0DgkzAwZdsJ27fg3kb0Yg/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/EMT21_water_column_measured_validated.csv\n",
      "Observatory_id PiEGetxo sheet_link https://docs.google.com/spreadsheets/d/1KGcishLP5eAn6ZCnixn5wZIez1Y9MgmYoch2s7aefKM/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='PiEGetxo' - sampling_strategy='water_column' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1KGcishLP5eAn6ZCnixn5wZIez1Y9MgmYoch2s7aefKM/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/PiEGetxo_water_column_measured_validated.csv\n",
      "Observatory_id Plenzia sheet_link https://docs.google.com/spreadsheets/d/13XeETquTvY6F_8Re25E0BAmSxLi0mThDUXYay3Eb5-o/edit?usp=sharing\n",
      "Observatory_id RFormosa sheet_link https://docs.google.com/spreadsheets/d/15f_yqrhvOr1MW-SHuv0JiKf0HRDy81uGZapJANxL54k/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='RFormosa' - sampling_strategy='water_column' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/15f_yqrhvOr1MW-SHuv0JiKf0HRDy81uGZapJANxL54k/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/RFormosa_water_column_measured_validated.csv\n",
      "Observatory_id OSD74 sheet_link https://docs.google.com/spreadsheets/d/1VATpzmhwmk8sYkhEjCbDUZHRiYuNGpSRefhXthZRbfQ/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='OSD74' - sampling_strategy='water_column' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1VATpzmhwmk8sYkhEjCbDUZHRiYuNGpSRefhXthZRbfQ/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/OSD74_water_column_measured_validated.csv\n",
      "Observatory_id AAOT sheet_link https://docs.google.com/spreadsheets/d/1hvLkBwiKTGTJDx19m_8e7qJ2lm9bwLLeVztMpxTLqnk/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='AAOT' - sampling_strategy='water_column' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1hvLkBwiKTGTJDx19m_8e7qJ2lm9bwLLeVztMpxTLqnk/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/AAOT_water_column_measured_validated.csv\n",
      "Observatory_id NRMCB sheet_link https://docs.google.com/spreadsheets/d/1F4AWv_seI-DQJ_Gp_N2ziwvGyjnc4KC92GGNXrJRek4/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='NRMCB' - sampling_strategy='water_column' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1F4AWv_seI-DQJ_Gp_N2ziwvGyjnc4KC92GGNXrJRek4/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/NRMCB_water_column_measured_validated.csv\n",
      "Observatory_id HCMR-1 sheet_link https://docs.google.com/spreadsheets/d/13DcVK2mzSxMJoFydSBaIMmj7Td1_JapEvcY2bmZTyLc/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='HCMR-1' - sampling_strategy='water_column' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/13DcVK2mzSxMJoFydSBaIMmj7Td1_JapEvcY2bmZTyLc/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/HCMR-1_water_column_measured_validated.csv\n",
      "Observatory_id IUIEilat sheet_link https://docs.google.com/spreadsheets/d/1qIwi1nZu4OBiriCzn1fEG1fyG8-3wSVvwzsMiH6JTwk/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='IUIEilat' - sampling_strategy='water_column' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1qIwi1nZu4OBiriCzn1fEG1fyG8-3wSVvwzsMiH6JTwk/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/IUIEilat_water_column_measured_validated.csv\n",
      "Observatory_id UMF sheet_link https://docs.google.com/spreadsheets/d/1-1VsUUbRtKselxu-y2BghRIrJdaf74SbmvK42ntvd20/edit#gid=2052143695\n",
      "\n",
      "\n",
      "Processing observatory_id='UMF' - sampling_strategy='water_column' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1-1VsUUbRtKselxu-y2BghRIrJdaf74SbmvK42ntvd20/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/UMF_water_column_measured_validated.csv\n",
      "Observatory_id LMO sheet_link https://docs.google.com/spreadsheets/d/1AvQMYcS0tdNMw6Er8zUarQg1a_wrshhnkTS6RuI1FJQ/edit#gid=2052143695\n",
      "\n",
      "\n",
      "Processing observatory_id='LMO' - sampling_strategy='water_column' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1AvQMYcS0tdNMw6Er8zUarQg1a_wrshhnkTS6RuI1FJQ/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Discarded 8 records leaving 110.\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/LMO_water_column_measured_validated.csv\n",
      "Observatory_id ESC68N sheet_link nan\n",
      "Observatory ESC68N lacks valid sheet URL for soft_sediment\n",
      "Observatory_id Bergen sheet_link nan\n",
      "Observatory Bergen lacks valid sheet URL for soft_sediment\n",
      "Observatory_id MBAL4 sheet_link nan\n",
      "Observatory MBAL4 lacks valid sheet URL for soft_sediment\n",
      "Observatory_id BPNS sheet_link https://docs.google.com/spreadsheets/d/1zc0bZdpl-Eoi35lI_5BGkElbscplyQRyNPLkSgeEyEQ/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='BPNS' - sampling_strategy='soft_sediment' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1zc0bZdpl-Eoi35lI_5BGkElbscplyQRyNPLkSgeEyEQ/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/BPNS_soft_sediment_measured_validated.csv\n",
      "Observatory_id ROSKOGO sheet_link https://docs.google.com/spreadsheets/d/1M0ytBWbUDP0YthGtTVtdIaX2cQUxF_uNIJBHIXREoX8/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='ROSKOGO' - sampling_strategy='soft_sediment' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1M0ytBWbUDP0YthGtTVtdIaX2cQUxF_uNIJBHIXREoX8/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/ROSKOGO_soft_sediment_measured_validated.csv\n",
      "Observatory_id VB sheet_link nan\n",
      "Observatory VB lacks valid sheet URL for soft_sediment\n",
      "Observatory_id OOB sheet_link https://docs.google.com/spreadsheets/d/1J469_ljfxM9NhOEyRmvWQNOXX7dujxBy1opU6ROmDv8/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='OOB' - sampling_strategy='soft_sediment' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1J469_ljfxM9NhOEyRmvWQNOXX7dujxBy1opU6ROmDv8/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/OOB_soft_sediment_measured_validated.csv\n",
      "Observatory_id EMT21 sheet_link https://docs.google.com/spreadsheets/d/1sBB0x6h-prnUHMcOfms_7qSqTn6zeGVriitTysceoT0/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='EMT21' - sampling_strategy='soft_sediment' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1sBB0x6h-prnUHMcOfms_7qSqTn6zeGVriitTysceoT0/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/EMT21_soft_sediment_measured_validated.csv\n",
      "Observatory_id PiEGetxo sheet_link nan\n",
      "Observatory PiEGetxo lacks valid sheet URL for soft_sediment\n",
      "Observatory_id Plenzia sheet_link nan\n",
      "Observatory Plenzia lacks valid sheet URL for soft_sediment\n",
      "Observatory_id RFormosa sheet_link https://docs.google.com/spreadsheets/d/17rqrZ-qrDP77SpoEyBFO17wewp98a-tLBWTlBAn6ZCc/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='RFormosa' - sampling_strategy='soft_sediment' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/17rqrZ-qrDP77SpoEyBFO17wewp98a-tLBWTlBAn6ZCc/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/RFormosa_soft_sediment_measured_validated.csv\n",
      "Observatory_id OSD74 sheet_link nan\n",
      "Observatory OSD74 lacks valid sheet URL for soft_sediment\n",
      "Observatory_id AAOT sheet_link nan\n",
      "Observatory AAOT lacks valid sheet URL for soft_sediment\n",
      "Observatory_id NRMCB sheet_link https://docs.google.com/spreadsheets/d/1a19q3w2sP7dae8HEzT5TdmfqhRQQd3GItFqiYxWJlxs/edit?usp=sharing\n",
      "\n",
      "\n",
      "Processing observatory_id='NRMCB' - sampling_strategy='soft_sediment' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1a19q3w2sP7dae8HEzT5TdmfqhRQQd3GItFqiYxWJlxs/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Discarded 1 records leaving 99.\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/NRMCB_soft_sediment_measured_validated.csv\n",
      "Observatory_id HCMR-1 sheet_link nan\n",
      "Observatory HCMR-1 lacks valid sheet URL for soft_sediment\n",
      "Observatory_id IUIEilat sheet_link nan\n",
      "Observatory IUIEilat lacks valid sheet URL for soft_sediment\n",
      "Observatory_id UMF sheet_link https://docs.google.com/spreadsheets/d/1DpQqo2i4q7RAM0sdm_z2pISa8SDBa9aYhWmUflFuUQs/edit?usp=drive_web&ouid=110054986816469121247\n",
      "\n",
      "\n",
      "Processing observatory_id='UMF' - sampling_strategy='soft_sediment' - sheet_type='measured'\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1DpQqo2i4q7RAM0sdm_z2pISa8SDBa9aYhWmUflFuUQs/gviz/tq?tqx=out:csv&sheet=measured\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/UMF_soft_sediment_measured_validated.csv\n",
      "Observatory_id LMO sheet_link nan\n",
      "Observatory LMO lacks valid sheet URL for soft_sediment\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%reload_ext autoreload\n",
    "\n",
    "import math\n",
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from validation_classes import measuredModel, samplingModel\n",
    "\n",
    "\n",
    "def get_sampling_or_measured_sheet(\n",
    "    sheet_type: str, sheet_link: str, format_type: str = \"json\"\n",
    ") -> pd.core.frame.DataFrame:\n",
    "    \"\"\"Returns a Pandas dataframe of the 'sampling' or 'measured' sheets\n",
    "    from the observatories' Google Sheets.\n",
    "\n",
    "    CSV has a problem with the word \"blank\" in the replicated field.\n",
    "    But none of the others work because of the header (not sure why only\n",
    "    CSV doesnt have the header.\n",
    "\n",
    "    TODO sort out the header parsing from the Google Sheet response.\n",
    "    \"\"\"\n",
    "    sampling_sheet_base = sheet_link.split(\"/edit\")[0]\n",
    "    if format_type == \"json\":\n",
    "        # should return json\n",
    "        sampling_sheet_suffix = f\"/gviz/tq?tqx=sheet={sheet_type}\"\n",
    "        sample_sheet_link = sampling_sheet_base + sampling_sheet_suffix\n",
    "        print(f\"Sample sheet link: {sample_sheet_link}\")\n",
    "        df = pd.read_json(sample_sheet_link)\n",
    "\n",
    "    elif format_type == \"csv\":\n",
    "        sampling_sheet_suffix = f\"/gviz/tq?tqx=out:csv&sheet={sheet_type}\"\n",
    "        sample_sheet_link = sampling_sheet_base + sampling_sheet_suffix\n",
    "        print(f\"Sample sheet link: {sample_sheet_link}\")\n",
    "\n",
    "        # Note that even if we force the replicate field to be a string, it doesnt\n",
    "        # recognise \"blank\" as a string, it's still None\n",
    "        # df = pd.read_csv(sample_sheet_link, encoding='utf-8', dtype={\"replicate\": str})\n",
    "        # If we force them to ints the NaNs become NA as noted above\n",
    "        # df = pd.read_csv(sample_sheet_link, encoding='utf-8', dtype={\"replicate\": int})\n",
    "\n",
    "        # Here we don't force ints with NaNs to \"Int64\", but\n",
    "        # let the later validator coerce floats and ints to string | None\n",
    "        df = pd.read_csv(sample_sheet_link, encoding=\"utf-8\")\n",
    "\n",
    "    elif format_type == \"excel\":\n",
    "        sampling_sheet_suffix = f\"/gviz/tq?tqx=out:tsv-xlsx&sheet={sheet_type}\"\n",
    "        sample_sheet_link = sampling_sheet_base + sampling_sheet_suffix\n",
    "        print(f\"Sample sheet link: {sample_sheet_link}\")\n",
    "        df = pd.read_excel(sample_sheet_link, engine=\"openpyxl\")\n",
    "    elif format_type == \"json\":\n",
    "        sampling_sheet_suffix = f\"/gviz/tq?tqx=out:json&sheet={sheet_type}\"\n",
    "        sample_sheet_link = sampling_sheet_base + sampling_sheet_suffix\n",
    "        print(f\"Sample sheet link: {sample_sheet_link}\")\n",
    "        df = pd.read_json(sample_sheet_link)\n",
    "    elif format_type == \"tsv\":\n",
    "        sampling_sheet_suffix = f\"/gviz/tq?tqx=out:tsv&sheet={sheet_type}\"\n",
    "        sample_sheet_link = sampling_sheet_base + sampling_sheet_suffix\n",
    "        print(f\"Sample sheet link: {sample_sheet_link}\")\n",
    "        df = pd.read_csv(sample_sheet_link, sep=\"\\t\")\n",
    "    else:\n",
    "        raise ValueError(f\"Unrecognised {format_type=}\")\n",
    "    return df\n",
    "\n",
    "\n",
    "def parse_sample_sheets(\n",
    "    sampling_strategy: str, sheet_type: str, addresses: list[str, str]\n",
    ") -> None:\n",
    "    for observatory in addresses:\n",
    "        observatory_id, sheet_link = observatory\n",
    "        print(f\"Observatory_id {observatory_id} sheet_link {sheet_link}\")\n",
    "\n",
    "        if not isinstance(sheet_link, str):\n",
    "            # print(f\"This is the sheet_link type {type(sheet_link)}\")\n",
    "            if isinstance(sheet_link, float):\n",
    "                if math.isnan(sheet_link):\n",
    "                    print(\n",
    "                        f\"Observatory {observatory_id} lacks valid sheet URL for {sampling_strategy}\"\n",
    "                    )\n",
    "                    continue\n",
    "            else:\n",
    "                raise ValueError(\n",
    "                    f\"Unknown link {sheet_link} to observatory {observatory_id}\"\n",
    "                )\n",
    "        else:\n",
    "            if observatory_id == \"Plenzia\":\n",
    "                continue  # Sheets not publically available\n",
    "\n",
    "            # if observatory_id not in [\"OSD74\", \"AAOT\"]:\n",
    "            #     continue\n",
    "            # if sampling_strategy != \"water_column\":\n",
    "            #     continue\n",
    "\n",
    "            print(\n",
    "                f\"\\n\\nProcessing {observatory_id=} - {sampling_strategy=} - {sheet_type=}\"\n",
    "            )\n",
    "            # Assuming either 'sampling' or 'measured' for sheet_type\n",
    "            df = get_sampling_or_measured_sheet(\n",
    "                sheet_type, sheet_link, format_type=\"csv\"\n",
    "            )\n",
    "\n",
    "            data_records_all = df.to_dict(orient=\"records\")\n",
    "\n",
    "            # Many sheets have partially filled rows\n",
    "            # The source_mat_id is auto-formatted and the PRIMARY_KEY\n",
    "            # Therefore filter records on source_mat_id\n",
    "            def filter_on_source_mat_id(d):\n",
    "                try:\n",
    "                    value = d[\"source_mat_id\"]\n",
    "                except KeyError as e:\n",
    "                    raise ValueError(\"Cannot find source_mat_id field\") from e\n",
    "                if isinstance(value, float):\n",
    "                    if math.isnan(value):\n",
    "                        return False\n",
    "                    else:  # not nan\n",
    "                        raise ValueError(\n",
    "                            f\"Unrecognised float value: {value} in source_mat_id\"\n",
    "                        )\n",
    "                elif value is None or len(value.split(\"_\")) < 4:\n",
    "                    return False\n",
    "\n",
    "                # VB_IMEV has \"à vérifier si les filets sont présents\" for source_mat_id_orig\n",
    "                # but valid source_mat_id\n",
    "                # https://github.com/emo-bon/observatory-profile/issues/37\n",
    "                try:\n",
    "                    value = d[\"source_mat_id_orig\"]\n",
    "                except KeyError:\n",
    "                    # Measured sheets dont have source_mat_id_orig\n",
    "                    return True\n",
    "                # Check the sampling sheets\n",
    "                return \"si les filets\" not in value\n",
    "\n",
    "            data_records_filtered = list(\n",
    "                filter(filter_on_source_mat_id, data_records_all)\n",
    "            )\n",
    "\n",
    "            if len(data_records_all) > len(data_records_filtered):\n",
    "                print(\n",
    "                    f\"Discarded {len(data_records_all) - len(data_records_filtered)} records leaving {len(data_records_filtered)}.\"\n",
    "                )\n",
    "\n",
    "            validator = validator_classes[sheet_type]\n",
    "            validated_rows = [\n",
    "                validator(**row).model_dump() for row in data_records_filtered\n",
    "            ]\n",
    "\n",
    "            save_dir = PROJECT_DIR / \"validated-data\" / \"logsheets\"\n",
    "            outfile_name = (\n",
    "                f\"{observatory_id}_{sampling_strategy}_{sheet_type}_validated.csv\"\n",
    "            )\n",
    "            outfile_name = save_dir / outfile_name\n",
    "            ndf = pd.DataFrame.from_records(validated_rows, index=\"source_mat_id\")\n",
    "\n",
    "            ndf.to_csv(outfile_name)\n",
    "            print(f\"Written {outfile_name}\")\n",
    "\n",
    "\n",
    "validator_classes = {\"sampling\": samplingModel, \"measured\": measuredModel}\n",
    "# Get list of observatory ids\n",
    "df = pd.read_csv(\n",
    "    PROJECT_DIR / \"validated-data\" / \"governance\" / \"observatories_validated.csv\"\n",
    ")\n",
    "observatory_ids = [id[0] for id in df[[\"observatory_id\"]].values.tolist()]\n",
    "# Get list of all URL links to sampling sheets\n",
    "# NB  you cant use a \"with\" closure here when reading the Pandas df\n",
    "df = pd.read_csv(\n",
    "    PROJECT_DIR / \"validated-data\" / \"governance\" / \"logsheets_validated.csv\"\n",
    ")\n",
    "water_column_sheet_addresses = df[[\"observatory_id\", \"water_column\"]].values.tolist()\n",
    "soft_sediment_sheet_addresses = df[[\"observatory_id\", \"soft_sediment\"]].values.tolist()\n",
    "\n",
    "parse_sample_sheets(\"water_column\", \"sampling\", water_column_sheet_addresses)\n",
    "parse_sample_sheets(\"soft_sediment\", \"sampling\", soft_sediment_sheet_addresses)\n",
    "parse_sample_sheets(\"water_column\", \"measured\", water_column_sheet_addresses)\n",
    "parse_sample_sheets(\"soft_sediment\", \"measured\", soft_sediment_sheet_addresses)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fa1b378-bbd4-4403-ae00-37c5b5ba8b82",
   "metadata": {},
   "source": [
    "# Validate the Observatory sheets the EMO-BON-Metadata Google Sheets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8a1f67c-e5a8-445d-ac90-1ad4bad88180",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "\n",
      "\n",
      "Processing observatory_id='ESC68N' - sampling_strategy='water_column' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/11_Eu0W1-sDiuzKx1cIl6YuxjRHmWezN6u9v3Ly8JZ3A/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/11_Eu0W1-sDiuzKx1cIl6YuxjRHmWezN6u9v3Ly8JZ3A/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 68.925895, 'longitude': 17.125619, 'geo_loc_name': 'Norway', 'loc_broad_ocean': 'Arctic Ocean', 'loc_broad_ocean_mrgid': 1906, 'loc_regional': 'Norwegian Sea', 'loc_regional_mrgid': 2353, 'loc_loc': 'Norwegian part of the Norwegian Sea', 'loc_loc_mrgid': 25242, 'env_broad_biome': 'marine biome [ENVO:00000447];marine pelagic biome [ENVO:01000023]', 'env_local': 'marine pelagic feature [ENVO:01000044];marine water body [ENVO:00001999];costal water body [ENVO:02000049];fjord [ENVO_00000039]', 'env_package': 'water', 'tot_depth_water_col': 18, 'organization': 'The Arctic University of Norway (UiT)', 'organization_country': 'Norway', 'organization_edmoid': 5354, 'obs_id': 'ESC68N', 'wa_id': 'ESC68N Wa', 'extra_site_info': 'Sampling site in the bottom end of a west norwegian fjord to monitoring the annual tide-driven flux of marine larva in and out of the fjord system. It has not been including in genomic biodiversity sampling campains in the past. The annual phytoplankton bloom peak is typically in late march. It is a recreational area with boat activity especially in the spring and summer period.', 'contact_name': 'Kim Praebel', 'contact_email': 'Kim.praebel@uit.no', 'contact_orcid': '0000-0002-0681-1854', 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51654', 'Unnamed: 25': nan, 'Unnamed: 26': nan, 'Unnamed: 27': nan}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/ESC68N_water_column_observatory_validated.csv\n",
      "\n",
      "\n",
      "Processing observatory_id='Bergen' - sampling_strategy='water_column' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1HuXHiUJICZrmCrJ4EZDyU5aSCMzDAc1cy_tne5YVPTg/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1HuXHiUJICZrmCrJ4EZDyU5aSCMzDAc1cy_tne5YVPTg/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 60.269671, 'longitude': 5.21957, 'geo_loc_name': 'Norway', 'loc_broad_ocean': 'North Atlantic Ocean', 'loc_broad_ocean_mrgid': 1912, 'loc_regional': 'North Sea', 'loc_regional_mrgid': 2350, 'loc_loc': 'Norwegian part of the North Sea', 'loc_loc_mrgid': 25233, 'env_broad_biome': 'marine biome [ENVO:00000447];marine pelagic biome [ENVO:01000023]', 'env_local': 'marine pelagic feature [ENVO:01000044];marine water body [ENVO:00001999];costal water body [ENVO:02000049];fjord [ENVO_00000039]', 'env_package': 'water', 'tot_depth_water_col': 18, 'organization': 'University of Bergen (UiB)', 'organization_country': 'Norway', 'organization_edmoid': 544, 'obs_id': 'Bergen', 'wa_id': 'Bergen Wa', 'extra_site_info': 'Sampling site in the bottom end of a west norwegian fjord to monitoring the annual tide-driven flux of marine larva in and out of the fjord system. It has not been including in genomic biodiversity sampling campains in the past. The annual phytoplankton bloom peak is typically in late march. It is a recreational area with boat activity especially in the spring and summer period.', 'contact_name': 'Henrik Glenner', 'contact_email': 'henrik.glenner@uib.no', 'contact_orcid': '0000-0002-8961-7319', 'ENA_accesion_number_umbrella': 'PRJEB51688', 'ENA_accesion_number_project': 'PRJEB50569'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/Bergen_water_column_observatory_validated.csv\n",
      "\n",
      "\n",
      "Processing observatory_id='MBAL4' - sampling_strategy='water_column' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1xfrqraPa0auQ1O-C9RUo68RhxrPCDWkVMCAUbj79AZI/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1xfrqraPa0auQ1O-C9RUo68RhxrPCDWkVMCAUbj79AZI/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 50.25, 'longitude': -4.217, 'geo_loc_name': 'United Kingdom', 'loc_broad_ocean': 'North Atlantic Ocean', 'loc_broad_ocean_mrgid': 1912, 'loc_regional': 'English Channel', 'loc_regional_mrgid': 2389, 'loc_loc': 'Western Channel', 'loc_loc_mrgid': 17527, 'env_broad_biome': 'marine biome [ENVO:00000447];marine pelagic biome [ENVO:01000023]', 'env_local': 'marine pelagic feature [ENVO:01000044];marine water body [ENVO:00001999];costal water body [ENVO:02000049]', 'env_package': 'water', 'tot_depth_water_col': 51, 'organization': 'Marine Biological Association', 'organization_country': 'United Kingdom', 'organization_edmoid': 45, 'obs_id': 'MBAL4', 'wa_id': 'MBAL4 Wa', 'extra_site_info': 'Station L4 is part of the Western Channel Observatory (WCO) off Plymouth. The long-term ecological site has been studied for over a century, with plankton sampled from viruses, bacteria and fungi to large zooplankton and fish larvae. The site is also utilised for adult fish surveys and eDNA-based assessments of marine biodiversity. Station L4 is typically seasonally stratified and sometimes impacted by estuarine waters from the adjacent rivers Tamar and Plym. In a typical year, a spring diatom bloom is followed by Phaeocystis and other major phytoplankton groups including dinoflagellates. The site has been subjected to a broad range of omics and molecular-based studies, including metabarcoding, metagenomics and single-cell approaches.', 'contact_name': 'Michael Cunliffe', 'contact_email': 'micnli@mba.ac.uk', 'contact_orcid': '0000-0002-6716-3555', 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51657', 'Unnamed: 25': nan, 'Unnamed: 26': nan, 'Unnamed: 27': nan}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/MBAL4_water_column_observatory_validated.csv\n",
      "\n",
      "\n",
      "Processing observatory_id='BPNS' - sampling_strategy='water_column' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1mEi4Bd2YR63WD0j54FQ6QkzcUw_As9Wilue9kaXO2DE/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1mEi4Bd2YR63WD0j54FQ6QkzcUw_As9Wilue9kaXO2DE/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 51.433331, 'longitude': 2.808331, 'geo_loc_name': 'Belgium', 'loc_broad_ocean': 'North Atlantic Ocean', 'loc_broad_ocean_mrgid': 1912, 'loc_regional': 'North Sea', 'loc_regional_mrgid': 2350, 'loc_loc': 'Belgian part of the North Sea', 'loc_loc_mrgid': 26567, 'env_broad_biome': 'marine biome [ENVO:00000447];marine pelagic biome [ENVO:01000023]', 'env_local': 'marine pelagic feature [ENVO:01000044];marine water body [ENVO:00001999];costal water body [ENVO:02000049]', 'env_package': 'water', 'tot_depth_water_col': 26, 'organization': 'Flanders Marine Institute (VLIZ);Ghent University;Royal Belgian Institute of Natural Sciences;Katholieke Universiteit Leuven', 'organization_country': 'Belgium', 'organization_edmoid': '422;449;432;2248', 'obs_id': 'BPNS', 'wa_id': 'BPNS Wa', 'extra_site_info': 'Station that has historic data going back to 1980’s and has monthly environmental measurements since 2001. Site participated in Ocean Sampling day activities for years 2012 2017 2018 2019 2020 and OSD12x. Parallel genomic biodiversity sampling in 2014 and monthly in 2016. The spring phytoplankton bloom (expressed as chlorophyll a) typically occurs between March to May. A spring diatom community is ‘followed’ by a bloom of Phaeocystis globosa and a summer diatom bloom. (Desmit. et al. 2015). A number of potentially harmful algal species occur such as Pseudo-nitzschia and several dinoflagellate genera (e.g. Prorocentrum Alexandrium..). (Nohe et al. 2019). BPNS is characterised by high antropogenic activity (Intense shipping traffic offshore windmill farms dredging etc.)', 'contact_name': 'Klaas Deneudt', 'contact_email': 'klaas.deneudt@vliz.be', 'contact_orcid': '0000-0002-8559-3508', 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51652'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/BPNS_water_column_observatory_validated.csv\n",
      "\n",
      "\n",
      "Processing observatory_id='ROSKOGO' - sampling_strategy='water_column' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1BCu2pbuIS4f-yrw5Kw4uzfO9v6ryZJBtSF5WBLWG_-E/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1BCu2pbuIS4f-yrw5Kw4uzfO9v6ryZJBtSF5WBLWG_-E/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 48.771667, 'longitude': -3.968333, 'geo_loc_name': 'France', 'loc_broad_ocean': 'North Atlantic Ocean', 'loc_broad_ocean_mrgid': 1912, 'loc_regional': 'English Channel', 'loc_regional_mrgid': 2389, 'loc_loc': 'French part of the English Channel', 'loc_loc_mrgid': 25214, 'env_broad_biome': 'marine biome [ENVO:00000447];marine pelagic biome [ENVO:01000023]', 'env_local': 'marine pelagic feature [ENVO:01000044];marine water body [ENVO:00001999];costal water body [ENVO:02000049]', 'env_package': 'water', 'tot_depth_water_col': 60, 'organization': 'Station Biologique de Roscoff', 'organization_country': 'France', 'organization_edmoid': 521, 'obs_id': 'ROSKOGO', 'wa_id': 'ROSKOGO Wa', 'extra_site_info': 'The SOMLIT-Astan time-series station is considered as representative of the French Western English Channel with an intense tidal mixing and winds preventing summer stratification. Long term monitoring of the hydrology (15 parameters) and phytoplankton (incuding DNA collection) has been initiated since 2000. These monitorings are operated in the frame of national monitoring programs : the SOMLIT - Service d’Observation en Milieu LITtoral – (http://somlit.epoc.u-bordeaux1.fr/) and the PHYTOBS - PHYtoplankton OBServatory (since 2018 - https://www.ir-ilico.fr/Les-reseaux-elementaires/Fiches-d-identite-des-reseaux-elementaires/PHYTOBS), respectively. Samples are collected bi-monthly during high neap tide at surface (1 m depth). Samples (0.2-3 µm and > 3µm) for metagenomics and metatranscriptomics are sampled since May 2018. Samples for viromes (DNA and RNA viruses) and for metagenomics and metatranscriptomics of micro- (> 20 µm) and mesoplankton (>200 µm) are sampled since June 2020. This site participated to all OSD events since 2014. No HAB events so far. The main annual phytoplankton bloom (dominated by the diatom Guinardia delicatula) occurs in May-June.', 'contact_name': 'Fabienne Rigaut-Jalabert', 'contact_email': 'fabienne.jalabert@sb-roscoff.fr', 'contact_orcid': nan, 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51662'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/ROSKOGO_water_column_observatory_validated.csv\n",
      "\n",
      "\n",
      "Processing observatory_id='VB' - sampling_strategy='water_column' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1kSlTcNfXaCQv7fIKVbBnt5PmRW3_H7Trryh4FJHyLkE/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1kSlTcNfXaCQv7fIKVbBnt5PmRW3_H7Trryh4FJHyLkE/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 43.683, 'longitude': 7.317, 'geo_loc_name': 'France', 'loc_broad_ocean': 'Mediterranean Sea - Western Basin', 'loc_broad_ocean_mrgid': 4279, 'loc_regional': 'Villefranche Bay', 'loc_regional_mrgid': 5367, 'loc_loc': 'Villefranche Bay - Point B', 'loc_loc_mrgid': 5368, 'env_broad_biome': 'marine biome [ENVO:00000447];marine pelagic biome [ENVO:01000023];mediterranean sea biome[ENVO_01000047]', 'env_local': 'marine pelagic feature [ENVO:01000044];marine water body [ENVO:00001999];costal water body [ENVO:02000049]', 'env_package': 'water', 'tot_depth_water_col': 80, 'organization': 'Institut de la Mer de Villefranche', 'organization_country': 'France', 'organization_edmoid': 5041, 'obs_id': 'VB', 'wa_id': 'VB Wa', 'extra_site_info': 'This sampling is coherent with the pointB long term series of villefranche sur mer (SOMLIT) which is operated since 1957 and is one of the longest time series in plankton. It includes a large and coherent environmental and biological survey (consultable here http://www.obs-vlfr.fr/data/view/) and have previously be part of OSD between 2012 and 2015. The long term sampling have also been used for various genomic sampling in the past (discontinued depending on projects). It is sometimes prone to HAB (with high recurrence of ostreopsis on summer) and is part of PHYTOBS surveys. It is prone to phytoplankton blooms (usually in March/April). It is complemented with high frequency CTD sampling and weelkly plankton tows. Further informations: http://rade.obs-vlfr.fr/Site.htm', 'contact_name': 'Cecile Fauvelot', 'contact_email': 'cecile.fauvelot@ird.fr', 'contact_orcid': 'http://orcid.org/0000-0003-0806-1222', 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51664'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/VB_water_column_observatory_validated.csv\n",
      "Observatory OOB lacks valid sheet URL for water_column\n",
      "\n",
      "\n",
      "Processing observatory_id='EMT21' - sampling_strategy='water_column' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1RutVpduwL1mEed5jC2zLbs0DgkzAwZdsJ27fg3kb0Yg/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1RutVpduwL1mEed5jC2zLbs0DgkzAwZdsJ27fg3kb0Yg/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 42.201944, 'longitude': -8.7985, 'geo_loc_name': 'Spain', 'loc_broad_ocean': 'Atlantic Ocean', 'loc_broad_ocean_mrgid': 1902, 'loc_regional': 'North Atlantic Ocean', 'loc_regional_mrgid': 1912, 'loc_loc': 'Vigo Seamount', 'loc_loc_mrgid': 4889, 'env_broad_biome': 'marine biome [ENVO:00000447];marine pelagic biome [ENVO:01000023]', 'env_local': 'marine pelagic feature [ENVO:01000044];marine water body [ENVO:00001999];costal water body [ENVO:02000049]', 'env_package': 'water', 'tot_depth_water_col': 5, 'organization': 'Toralla Marine Science Station - Centre of Marine Research, University of Vigo', 'organization_country': 'Spain', 'organization_edmoid': 2163, 'obs_id': 'EMT21', 'wa_id': 'EMT21 Wa', 'extra_site_info': 'This site is located at the marina of the ECIMAT marine station, where there is a real-time monitoring station for environmental variables (with a 8 years-long time series). This point is in the same estuary where OSD has been conducted since 2018. Harmful algal blooms are periodically detected in the region during the summer months and it is located in an upwelling region, so phytoplankton blooms related to the upwelling events are frequent from March to September.', 'contact_name': 'Jose Gonzalez; Jesús Troncoso', 'contact_email': 'josegonzález@uvigo.es; troncoso@uvigo.es', 'contact_orcid': '0000-0003-1576-9819', 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51653'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/EMT21_water_column_observatory_validated.csv\n",
      "\n",
      "\n",
      "Processing observatory_id='PiEGetxo' - sampling_strategy='water_column' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1KGcishLP5eAn6ZCnixn5wZIez1Y9MgmYoch2s7aefKM/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1KGcishLP5eAn6ZCnixn5wZIez1Y9MgmYoch2s7aefKM/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 43.338583, 'longitude': -3.014639, 'geo_loc_name': 'Spain', 'loc_broad_ocean': 'North Atlantic Ocean', 'loc_broad_ocean_mrgid': 1912, 'loc_regional': 'Bay of Biscay', 'loc_regional_mrgid': 2359, 'loc_loc': 'Abra de Bilbao', 'loc_loc_mrgid': 17355, 'env_broad_biome': 'marine biome [ENVO:00000447];estuarine biome [ENVO:01000020]', 'env_local': 'marine pelagic feature [ENVO:01000044];marine water body [ENVO:00001999];harbor[ENVO:00000463]', 'env_package': 'water', 'tot_depth_water_col': 10, 'organization': 'Research Centre for Experimental Marine Biology and Biotechnology, Plentzia Marine Station', 'organization_country': 'Spain', 'organization_edmoid': 2167, 'obs_id': 'PiEGetxo', 'wa_id': 'PiEGetxo Wa', 'extra_site_info': 'The selected sampling site is a Marina, close to the main port of Bilbao (Basque Coast, Southern Bay of Biscay). The same site has been sampled in previous years for OSD2019; and OSD2020x12 and for ARMs (2019 and 2020). There are occasional red algal blooms in June and phytoplakton blooms occur normaly from March to April. Monitoring activities in the area include: chemical burden metals and PAHs (sediment and molluscs), acidification, zoo- and phytoplankton diversity, biological effects monitoring\\n', 'contact_name': 'Oihane Diaz De Cerio Arruabarrena', 'contact_email': 'oihane.diazdecerio@ehu.eus', 'contact_orcid': '0000-0002-5605-8434', 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51660'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/PiEGetxo_water_column_observatory_validated.csv\n",
      "\n",
      "\n",
      "Processing observatory_id='RFormosa' - sampling_strategy='water_column' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/15f_yqrhvOr1MW-SHuv0JiKf0HRDy81uGZapJANxL54k/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/15f_yqrhvOr1MW-SHuv0JiKf0HRDy81uGZapJANxL54k/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 37.005639, 'longitude': -7.96925, 'geo_loc_name': 'Portugal', 'loc_broad_ocean': 'Atlantic Ocean', 'loc_broad_ocean_mrgid': 1902, 'loc_regional': 'North Atlantic Ocean', 'loc_regional_mrgid': 1912, 'loc_loc': 'Ria Formosa', 'loc_loc_mrgid': 29473, 'env_broad_biome': 'marine biome [ENVO:00000447]', 'env_local': 'lagoon [ENVO_00000038];marine water body [ENVO:00001999]', 'env_package': 'water', 'tot_depth_water_col': 1, 'organization': 'Centre of Marine Sciences (CCMAR)', 'organization_country': 'Portugal', 'organization_edmoid': 2516, 'obs_id': 'RFormosa', 'wa_id': 'RFormosa Wa', 'extra_site_info': 'Yes, it’s OSD sampling site ID OSD81, annually sampled in OSD2014, OSD2015, OSD2016, OSD2018, OSD2019, and 12x monthly series of OSD2020. Is near the UALG experimental station Ramalhete. The site is a channel of a lagunar system and sampling always take place with out going current (lowering tide).', 'contact_name': 'Bruno Louro', 'contact_email': 'blouro@ualg.pt', 'contact_orcid': '0000-0001-8164-581X', 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51661'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/RFormosa_water_column_observatory_validated.csv\n",
      "\n",
      "\n",
      "Processing observatory_id='OSD74' - sampling_strategy='water_column' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1VATpzmhwmk8sYkhEjCbDUZHRiYuNGpSRefhXthZRbfQ/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1VATpzmhwmk8sYkhEjCbDUZHRiYuNGpSRefhXthZRbfQ/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 41.146528, 'longitude': -8.666639, 'geo_loc_name': 'Portugal', 'loc_broad_ocean': 'Atlantic Ocean', 'loc_broad_ocean_mrgid': 1902, 'loc_regional': 'North Atlantic Ocean', 'loc_regional_mrgid': 1912, 'loc_loc': 'Porto Valley', 'loc_loc_mrgid': 4888, 'env_broad_biome': 'marine biome [ENVO:00000447];marine pelagic biome [ENVO:01000023];estuarine biome [ENVO:01000020]', 'env_local': 'marine pelagic feature [ENVO:01000044];marine water body [ENVO:00001999];costal water body [ENVO:02000049]', 'env_package': 'water', 'tot_depth_water_col': 1, 'organization': 'Interdisciplinary Centre of Marine and Environmental Research', 'organization_country': 'Portugal', 'organization_edmoid': 1626, 'obs_id': 'OSD74', 'wa_id': 'OSD74 Wa', 'extra_site_info': 'This site participated in OSD2014, OSD2015 and OSD2016 during MicroB3 project and in OSD 2018 during ASSEMBLE Plus. This is a station located in the mouth of the Douro Estuary, with annual phytoplankton blooms occurring during spring. In this area there are several recreation and fishing activities.', 'contact_name': 'Catarina Magalhães', 'contact_email': 'cmagalhaes@ciimar.up.pt', 'contact_orcid': '0000-0001-9576-2398', 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51659'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/OSD74_water_column_observatory_validated.csv\n",
      "\n",
      "\n",
      "Processing observatory_id='AAOT' - sampling_strategy='water_column' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1hvLkBwiKTGTJDx19m_8e7qJ2lm9bwLLeVztMpxTLqnk/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1hvLkBwiKTGTJDx19m_8e7qJ2lm9bwLLeVztMpxTLqnk/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 45.314167, 'longitude': 12.508333, 'geo_loc_name': 'Italy', 'loc_broad_ocean': 'Mediterranean Sea - Eastern Basin', 'loc_broad_ocean_mrgid': 4280, 'loc_regional': 'Adriatic Sea', 'loc_regional_mrgid': 3314, 'loc_loc': 'Gulf of Venice', 'loc_loc_mrgid': 3610, 'env_broad_biome': 'marine biome [ENVO:00000447];marine pelagic biome [ENVO:01000023];mediterranean sea biome[ENVO_01000047]', 'env_local': 'marine pelagic feature [ENVO:01000044];marine water body [ENVO:00001999];costal water body [ENVO:02000049]', 'env_package': 'water', 'tot_depth_water_col': 17, 'organization': 'National Research Council Institute of Marine Science (CNR-ISMAR)', 'organization_country': 'Italy', 'organization_edmoid': 227, 'obs_id': 'AAOT', 'wa_id': 'AAOT Wa', 'extra_site_info': 'The site is part of the LTER network and it’s sampled at two levels every month for biogeochemical parameters (since 2010). Hydrological parameters area acquired automatically every 10’ at three levels. We sampled the site for OSD events for 2014 2015 and 2016 years. We monitor harmul species (HAB) and for the area DSP events are quite common due to the presence of Diophysis spp. We usually observe late winter and spring blooms.', 'contact_name': 'Mauro Bastianini', 'contact_email': 'mauro.bastianini@ismar.cnr.it', 'contact_orcid': '0000-0001-6758-4192', 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB50566'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/AAOT_water_column_observatory_validated.csv\n",
      "\n",
      "\n",
      "Processing observatory_id='NRMCB' - sampling_strategy='water_column' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1F4AWv_seI-DQJ_Gp_N2ziwvGyjnc4KC92GGNXrJRek4/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1F4AWv_seI-DQJ_Gp_N2ziwvGyjnc4KC92GGNXrJRek4/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 40.800139, 'longitude': 14.25, 'geo_loc_name': 'Italy', 'loc_broad_ocean': 'Mediterranean Sea - Western Basin', 'loc_broad_ocean_mrgid': 4279, 'loc_regional': 'Tyrrhenian Sea', 'loc_regional_mrgid': 3386, 'loc_loc': 'Naples Gulf', 'loc_loc_mrgid': 3558, 'env_broad_biome': 'marine biome [ENVO:00000447];marine pelagic biome [ENVO:01000023];mediterranean sea biome[ENVO_01000047]', 'env_local': 'marine pelagic feature [ENVO:01000044];marine water body [ENVO:00001999];costal water body [ENVO:02000049]', 'env_package': 'water', 'tot_depth_water_col': 75, 'organization': 'Stazione Zoologica Anton Dohrn', 'organization_country': 'Italy', 'organization_edmoid': 237, 'obs_id': 'NRMCB', 'wa_id': 'NRMCB Wa', 'extra_site_info': 'This is a LTER site, also participating to the OSD initiative since its beginning and currently running monthly. The site is object of monthly genomic diversity sampling since 2019 (www.nerea-observatory.org) and sporadically also in the past. We observe two recurrent phytoplankton blooms, in the spring and in the fall. Pico/Micro phytoplankton blooms are also observed in other periods, such as Synechochoccus in July and picoeukaryotes all year around, generally after rain events. Heterotrophic bacteria also show sporadic peaks after rains.The NEREA Observatory also has 2 other sampling points, visited seasonally, at the Sarno river and close to the island of Capri, at a deep site (800 m). The LTER site encompasses the whole Gulf of Naples and it includes a benthic site in the island of Ischia. 2 miles off mussel farms. Close to shellfish farms', 'contact_name': 'Raffaella Casotti', 'contact_email': 'raffaella.casotti@szn.it', 'contact_orcid': '0000-0002-9876-4601', 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51658'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/NRMCB_water_column_observatory_validated.csv\n",
      "\n",
      "\n",
      "Processing observatory_id='HCMR-1' - sampling_strategy='water_column' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/13DcVK2mzSxMJoFydSBaIMmj7Td1_JapEvcY2bmZTyLc/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/13DcVK2mzSxMJoFydSBaIMmj7Td1_JapEvcY2bmZTyLc/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 35.346621, 'longitude': 25.278761, 'geo_loc_name': 'Greece', 'loc_broad_ocean': 'Mediterranean Sea - Eastern Basin', 'loc_broad_ocean_mrgid': 4280, 'loc_regional': 'Aegean Sea', 'loc_regional_mrgid': 3315, 'loc_loc': 'Crete Sea', 'loc_loc_mrgid': 3339, 'env_broad_biome': 'marine biome [ENVO:00000447];marine pelagic biome [ENVO:01000023];mediterranean sea biome[ENVO_01000047]', 'env_local': 'marine pelagic feature [ENVO:01000044];marine water body [ENVO:00001999];costal water body [ENVO:02000049]', 'env_package': 'water', 'tot_depth_water_col': 21, 'organization': 'Institute of Marine Biology Biotechnology and Aquaculture (IMBBC) Hellenic Centre for Marine Research (HCMR)', 'organization_country': 'Greece', 'organization_edmoid': 141, 'obs_id': 'HCMR-1', 'wa_id': 'HCMR-1 Wa', 'extra_site_info': 'OSD100 for OSD campaign is participating annually\\nSea bottom: mainly coarse sand with macroalgal cover (Caulerpa prolifera) seagrass patches and escarpment (Posidonia oceanica) occasional small reefs. \\nTemperature data available from Seafloor Observatory at the maximum depth (sampling rate 5 minutes) and by temperature loggers deployed along the water column (18 16 14 12 10 8 m depth at hourly time intervals). \\nThe Seafloor Observatory at the UBPC comprises a CTD a Fluorometer and Acoustic Doppler Profiler (ADP) which are currently unavailable due to maintenance and scheduled to be redeployed by the end of 2021.\\nAnnual phytoplankton blooms occur in spring (March to May).\\nSince UBPC is an experimental platform research activities are taking place on the seafloor including experimental sponge cultivation and deployment and monitoring of artificial reefs deployment and monitoring of ARMS. Artisanal fisheries common in the surroundings but prohibited within the limits of the experimental facility. \\n', 'contact_name': 'Melanthia Stavroulaki', 'contact_email': 'mstavroulaki@hcmr.gr', 'contact_orcid': '0000-0003-4392-7159', 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51655'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/HCMR-1_water_column_observatory_validated.csv\n",
      "\n",
      "\n",
      "Processing observatory_id='IUIEilat' - sampling_strategy='water_column' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1qIwi1nZu4OBiriCzn1fEG1fyG8-3wSVvwzsMiH6JTwk/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1qIwi1nZu4OBiriCzn1fEG1fyG8-3wSVvwzsMiH6JTwk/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 29.5, 'longitude': 34.916667, 'geo_loc_name': 'Israel', 'loc_broad_ocean': 'Indian Ocean', 'loc_broad_ocean_mrgid': 1904, 'loc_regional': 'Gulf of Eilat', 'loc_regional_mrgid': 4263, 'loc_loc': 'Gulf of Eilat', 'loc_loc_mrgid': 4263, 'env_broad_biome': 'marine biome [ENVO:00000447];marine pelagic biome [ENVO:01000023]', 'env_local': 'marine pelagic feature [ENVO:01000044];marine water body [ENVO:00001999];costal water body [ENVO:02000049]', 'env_package': 'water', 'tot_depth_water_col': 700, 'organization': 'Interuniversity Institute for Marine Sciences in Eilat', 'organization_country': 'Israel', 'organization_edmoid': nan, 'obs_id': 'IUIEilat', 'wa_id': 'IUIEilat Wa', 'extra_site_info': 'There is a times series (National Monitoring Program- Israel). Yes we did sampling for OSD for 2018 2019 2020. No – not included in any genomic biodiversity sampling in the past. No HAB. Phytoplankton bloom is in March/April. Many other scientific activities take place in the site of sampling which is located offshore the Interuniversity Institue for Marine Sciences in Eilat. More info on the sampling site: https://www.sciencedirect.com/science/article/abs/pii/S0079661117301659', 'contact_name': 'Miguel Frada', 'contact_email': 'miguel.frada@mail.huji.ac.il', 'contact_orcid': nan, 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51656'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/IUIEilat_water_column_observatory_validated.csv\n",
      "\n",
      "\n",
      "Processing observatory_id='UMF' - sampling_strategy='water_column' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1-1VsUUbRtKselxu-y2BghRIrJdaf74SbmvK42ntvd20/edit#gid=2052143695\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1-1VsUUbRtKselxu-y2BghRIrJdaf74SbmvK42ntvd20/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 63.525, 'longitude': 19.808167, 'geo_loc_name': 'Sweden', 'loc_broad_ocean': 'Baltic Sea', 'loc_broad_ocean_mrgid': 2401, 'loc_regional': 'Gulf of Bothnia', 'loc_regional_mrgid': 2402, 'loc_loc': 'Bothnian Sea', 'loc_loc_mrgid': 19384, 'env_broad_biome': 'marine biome [ENVO:00000447];', 'env_local': 'marine water body [ENVO:00001999]', 'env_package': 'water', 'tot_depth_water_col': 23, 'organization': 'Umea Marine Sciences Centre  - Umea University', 'organization_country': 'Sweden', 'organization_edmoid': 622, 'obs_id': 'UMF', 'wa_id': 'UMF Wa', 'extra_site_info': 'This station (B7) is part of Sweden’s national monitoring program and thus there is an environmental time series since 2001 with hydrographic, biologic, and chemical parameters included. Benthic soft bottom fauna is also monitored in the area for the Swedish national monitoring program. A buoy is located near this station that records weather and water parameters every hour when it is deployed. The spring bloom occurs from April to May every year and there can often be a diatom bloom in September or October. Cyanobacteria blooms commonly occur here in late summer and autumn. Genomic samples have previously been taken at this station by a professor at Umeå University for a year.', 'contact_name': 'Annie Marie Cox', 'contact_email': 'annie.cox@umu.se', 'contact_orcid': nan, 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB57914'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/UMF_water_column_observatory_validated.csv\n",
      "\n",
      "\n",
      "Processing observatory_id='LMO' - sampling_strategy='water_column' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1AvQMYcS0tdNMw6Er8zUarQg1a_wrshhnkTS6RuI1FJQ/edit#gid=2052143695\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1AvQMYcS0tdNMw6Er8zUarQg1a_wrshhnkTS6RuI1FJQ/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 56.916904, 'longitude': 17.050178, 'geo_loc_name': 'Sweden', 'loc_broad_ocean': 'Baltic Sea', 'loc_broad_ocean_mrgid': 2401, 'loc_regional': 'Swedish part of the Baltic Sea', 'loc_regional_mrgid': 25678, 'loc_loc': 'Swedish part of the Baltic Sea', 'loc_loc_mrgid': 25678, 'env_broad_biome': 'marine biome [ENVO:00000447]', 'env_local': 'marine water body [ENVO:00001999]', 'env_package': 'water', 'tot_depth_water_col': 41, 'organization': 'Linnaeus University', 'organization_country': 'Sweden', 'organization_edmoid': 4812, 'obs_id': 'LMO', 'wa_id': 'LMO Wa', 'extra_site_info': 'At the LMO station we have been sampling (essentially twice monthly) for a diverse set of microbial evology variables since March 2011 (now at 361 sampling occasions). This includes measures of Bacterial abundance, Bacterial production, Inorganic nutrients and DOC, Chl a, phytoplankton composition (microscopy). We have also done 16S (all years) and 18S (several years) gene amplicon sequencing and metagenomics and metatranscriptomics analyses (5 years in total). Site has not participated in Ocean Sampling Day. Harmful algal blooms occur nearly every summer in the form of cyanobacterial blooms (especially Nodularia spumigena). There are three main periods of elevated Chl a: spring bloom in April (diatoms and dinoflagellates), summer bloom in July/August (cyanobacteria), and autumn bloom in Sept/Oct (diatoms). In the nearby coastal bays we investigate the influence of climate warming on microbial communities in water and sediments.', 'contact_name': 'Jarone Pinhassi', 'contact_email': 'Jarone.pinhassi@lnu.se', 'contact_orcid': '0000-0002-6405-1347', 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB64705'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/LMO_water_column_observatory_validated.csv\n",
      "Observatory ESC68N lacks valid sheet URL for soft_sediment\n",
      "Observatory Bergen lacks valid sheet URL for soft_sediment\n",
      "Observatory MBAL4 lacks valid sheet URL for soft_sediment\n",
      "\n",
      "\n",
      "Processing observatory_id='BPNS' - sampling_strategy='soft_sediment' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1zc0bZdpl-Eoi35lI_5BGkElbscplyQRyNPLkSgeEyEQ/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1zc0bZdpl-Eoi35lI_5BGkElbscplyQRyNPLkSgeEyEQ/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 51.433331, 'longitude': 2.80833, 'geo_loc_name': 'Belgium', 'loc_broad_ocean': 'North Atlantic Ocean', 'loc_broad_ocean_mrgid': 1912, 'loc_regional': 'North Sea', 'loc_regional_mrgid': 2350, 'loc_loc': 'Belgian part of the North Sea', 'loc_loc_mrgid': 26567, 'env_broad_biome': 'marine biome [ENVO:00000447];marine benthic biome[ENVO_01000024]', 'env_local': 'shallow_marine_sediment_[ENVO:03000034]', 'env_package': 'sediment', 'tot_depth_water_col': 26, 'organization': 'Flanders Marine Institute (VLIZ); Ghent University; Royal Belgian Institute of Natural Sciences; Katholieke Universiteit Leuven', 'organization_country': 'Belgium', 'organization_edmoid': '422; 449; 432; 2248', 'obs_id': 'BPNS', 'so_id': 'BPNS So', 'extra_site_info': 'Station that has historic data going back to 1980’s and has monthly environmental measurements since 2001. Site participated in Ocean Sampling day activities for years 2012; 2017; 2018; 2019; 2020 and OSD12x. Parallel genomic biodiversity sampling in 2014 and monthly in 2016. The spring phytoplankton bloom (expressed as chlorophyll a) typically occurs between March to May. A spring diatom community is ‘followed’ by a bloom of Phaeocystis globosa and a summer diatom bloom. (Desmit. et al. 2015). A number of potentially harmful algal species occur such as Pseudo-nitzschia and several dinoflagellate genera (e.g. Prorocentrum; Alexandrium..). (Nohe et al. 2019). BPNS is characterised by high antropogenic activity (Intense shipping traffic; offshore windmill farms; dredging; etc.)', 'contact_name': 'Klaas Deneudt', 'contact_email': 'klaas.deneudt@vliz.be', 'contact_orcid': '0000-0002-8559-3508', 'sediment_type': nan, 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51652'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/BPNS_soft_sediment_observatory_validated.csv\n",
      "\n",
      "\n",
      "Processing observatory_id='ROSKOGO' - sampling_strategy='soft_sediment' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1M0ytBWbUDP0YthGtTVtdIaX2cQUxF_uNIJBHIXREoX8/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1M0ytBWbUDP0YthGtTVtdIaX2cQUxF_uNIJBHIXREoX8/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 48.708333, 'longitude': -3.866, 'geo_loc_name': 'France', 'loc_broad_ocean': 'North Atlantic Ocean', 'loc_broad_ocean_mrgid': 1912, 'loc_regional': 'English Channel', 'loc_regional_mrgid': 2389, 'loc_loc': 'French part of the English Channel', 'loc_loc_mrgid': 25214, 'env_broad_biome': 'marine biome [ENVO:00000447];marine benthic biome[ENVO_01000024]', 'env_local': 'shallow_marine_sediment_[ENVO:03000034]', 'env_package': 'sediment', 'tot_depth_water_col': 18, 'organization': 'Station Biologique de Roscoff', 'organization_country': 'France', 'organization_edmoid': 521, 'obs_id': 'ROSKOGO', 'so_id': 'ROSKOGO So', 'extra_site_info': 'The sampling site (i.e. “Pierre Noire” station) is monitored for macrobenthic communities since 1977 twice a year (March and October). The site harbours an infralittoral fine sand community dominated by amphipods from the genus Ampelisca, polychaetes (Spiophanes bombyx, Nepthys spp., Hyalinoecia bilineata, Marphysa bellii) and bivalves (Abra alba, Timoclea ovata). In parallel to biological samples, temperature, salinity and sediment grain size are measured at this site. Hydrological variables (temperature, salinity, nutrients, chl.a, MES,…..) are monitored since 1997 at two nearby stations (SOMLIT-Estacade and SOMLIT-Astan stations) in the framework of the national service SOMLIT. The main annual phytoplankton bloom, dominated by diatoms, occurs in May-June. The sampling site was polluted by hydrocarbons following the Amoco-Cadiz oil spill which occurred in March 1978, one year after the beginning of the monitoring program. Significant hydrocarbon concentrations were measured in the sediment from 1978 to 1981. No major pollution has been reported in the bay since 1978. An aggregate dredging site is present at 2 nautical miles in the orth of the sampling site. The sampling site is located in a Natura 2000 site.', 'contact_name': 'Fabienne Rigaut-Jalabert', 'contact_email': 'fabienne.jalabert@sb-roscoff.fr', 'contact_orcid': nan, 'sediment_type': 'terrigenous', 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51662'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/ROSKOGO_soft_sediment_observatory_validated.csv\n",
      "Observatory VB lacks valid sheet URL for soft_sediment\n",
      "\n",
      "\n",
      "Processing observatory_id='OOB' - sampling_strategy='soft_sediment' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1J469_ljfxM9NhOEyRmvWQNOXX7dujxBy1opU6ROmDv8/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1J469_ljfxM9NhOEyRmvWQNOXX7dujxBy1opU6ROmDv8/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 42.484167, 'longitude': 3.135278, 'geo_loc_name': 'France', 'loc_broad_ocean': 'Mediterranean Sea - Western Basin', 'loc_broad_ocean_mrgid': 4279, 'loc_regional': 'Gulf of Lion', 'loc_regional_mrgid': 3452, 'loc_loc': 'Bay of Banyuls-sur-Mer', 'loc_loc_mrgid': 5207, 'env_broad_biome': 'marine biome [ENVO:00000447];marine benthic biome[ENVO_01000024]', 'env_local': 'shallow_marine_sediment_[ENVO:03000034]', 'env_package': 'sediment', 'tot_depth_water_col': 27, 'organization': 'Observatoire Océanologique de Banyuls', 'organization_country': 'France', 'organization_edmoid': 1015, 'obs_id': 'OOB', 'so_id': 'OOB So', 'extra_site_info': 'Part of the SOMLIT national network of observation, sampled for planktonic DNA for more than 10 years, participate to the Ocean Sampling Day since it started, blooms in late winter, situated close to a marine protected area.', 'contact_name': 'Pierre GALAND', 'contact_email': 'pierre.galand@obs-banyuls.fr', 'contact_orcid': nan, 'sediment_type': nan, 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51665', 'Unnamed: 26': nan, 'Unnamed: 27': nan}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/OOB_soft_sediment_observatory_validated.csv\n",
      "\n",
      "\n",
      "Processing observatory_id='EMT21' - sampling_strategy='soft_sediment' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1sBB0x6h-prnUHMcOfms_7qSqTn6zeGVriitTysceoT0/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1sBB0x6h-prnUHMcOfms_7qSqTn6zeGVriitTysceoT0/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 42.201944, 'longitude': -8.7985, 'geo_loc_name': 'Spain', 'loc_broad_ocean': 'Atlantic Ocean', 'loc_broad_ocean_mrgid': 1902, 'loc_regional': 'North Atlantic Ocean', 'loc_regional_mrgid': 1912, 'loc_loc': 'Vigo Seamount', 'loc_loc_mrgid': 4889, 'env_broad_biome': 'marine biome [ENVO:00000447];marine benthic biome[ENVO_01000024]', 'env_local': 'shallow_marine_sediment_[ENVO:03000034]', 'env_package': 'sediment', 'tot_depth_water_col': 5, 'organization': 'Estación de Ciencias Mariñas de Toralla - Centre of Marine Research, University of Vigo', 'organization_country': 'Spain', 'organization_edmoid': 2163, 'obs_id': 'EMT21', 'so_id': 'EMT21 So', 'extra_site_info': 'This site is located at the marina of the ECIMAT marine station, where there is a real-time monitoring station for environmental variables (with a 8 years-long time series). It is located in an upwelling region, so phytoplankton blooms related to the upwelling events are frequent from March to September.', 'contact_name': 'Jose Gonzalez, Jesús Troncoso', 'contact_email': 'josegonzález@uvigo.es, troncoso@uvigo.es', 'contact_orcid': '0000-0002-0202-8256', 'sediment_type': nan, 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51653'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/EMT21_soft_sediment_observatory_validated.csv\n",
      "Observatory PiEGetxo lacks valid sheet URL for soft_sediment\n",
      "Observatory Plenzia lacks valid sheet URL for soft_sediment\n",
      "\n",
      "\n",
      "Processing observatory_id='RFormosa' - sampling_strategy='soft_sediment' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/17rqrZ-qrDP77SpoEyBFO17wewp98a-tLBWTlBAn6ZCc/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/17rqrZ-qrDP77SpoEyBFO17wewp98a-tLBWTlBAn6ZCc/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 37.005639, 'longitude': -7.96925, 'geo_loc_name': 'Portugal', 'loc_broad_ocean': 'Atlantic Ocean', 'loc_broad_ocean_mrgid': 1902, 'loc_regional': 'North Atlantic Ocean', 'loc_regional_mrgid': 1912, 'loc_loc': 'Ria Formosa', 'loc_loc_mrgid': 29473, 'env_broad_biome': 'marine biome [ENVO:00000447];marine benthic biome[ENVO_01000024]', 'env_local': 'lagoon [ENVO_00000038];shallow_marine_sediment [ENVO:03000034]', 'env_package': 'sediment', 'tot_depth_water_col': 0, 'organization': 'Centre of Marine Sciences (CCMAR)', 'organization_country': 'Portugal', 'organization_edmoid': 2516, 'obs_id': 'RFormosa', 'so_id': 'RFormosa So', 'extra_site_info': 'Is silt mud, no metagenomics/metabarcoding time series, except the same site OSD sampling on the water column. Nearby the UALG experimental research station Ramalhete.', 'contact_name': 'Bruno Louro', 'contact_email': 'blouro@ualg.pt', 'contact_orcid': '0000-0001-8164-581X', 'sediment_type': 'biogenous', 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51661'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/RFormosa_soft_sediment_observatory_validated.csv\n",
      "Observatory OSD74 lacks valid sheet URL for soft_sediment\n",
      "Observatory AAOT lacks valid sheet URL for soft_sediment\n",
      "\n",
      "\n",
      "Processing observatory_id='NRMCB' - sampling_strategy='soft_sediment' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1a19q3w2sP7dae8HEzT5TdmfqhRQQd3GItFqiYxWJlxs/edit?usp=sharing\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1a19q3w2sP7dae8HEzT5TdmfqhRQQd3GItFqiYxWJlxs/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 40.800139, 'longitude': 14.25, 'geo_loc_name': 'Italy', 'loc_broad_ocean': 'Mediterranean Sea - Western Basin', 'loc_broad_ocean_mrgid': 4279, 'loc_regional': 'Tyrrhenian Sea', 'loc_regional_mrgid': 3386, 'loc_loc': 'Naples Gulf', 'loc_loc_mrgid': 3558, 'env_broad_biome': 'marine biome [ENVO:00000447];marine benthic biome[ENVO_01000024]', 'env_local': 'shallow_marine_sediment_[ENVO:03000034]', 'env_package': 'sediment', 'tot_depth_water_col': 75, 'organization': 'Stazione Zoologica Anton Dohrn', 'organization_country': 'Italy', 'organization_edmoid': 237, 'obs_id': 'NRMCB', 'so_id': 'NRMCB So', 'extra_site_info': 'The site has been sampled in the past (2010-2012). Water column info: This is a LTER site, also participating to the OSD initiative since its beginning and currently running monthly. The site is object of monthly genomic diversity sampling since 2019 (www.nerea-observatory.org) and sporadically also in the past. We observe two recurrent phytoplankton blooms, in the spring and in the fall. Pico/Micro phytoplankton blooms are also observed in other periods, such as Synechochoccus in July and picoeukaryotes all year around, generally after rain events. Heterotrophic bacteria also show sporadic peaks after rains.The NEREA Observatory also has 2 other sampling points, visited seasonally, at the Sarno river and close to the island of Capri, at a deep site (800 m). The LTER site encompasses the whole Gulf of Naples and it includes a benthic site in the island of Ischia. the site is 2 miles off a mussel farm. Close to shellfish farms', 'contact_name': 'Raffaella Casotti', 'contact_email': 'raffaella.casotti@szn.it', 'contact_orcid': '0000-0002-9876-4601', 'sediment_type': 'terrigenous', 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB51658'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/NRMCB_soft_sediment_observatory_validated.csv\n",
      "Observatory HCMR-1 lacks valid sheet URL for soft_sediment\n",
      "Observatory IUIEilat lacks valid sheet URL for soft_sediment\n",
      "\n",
      "\n",
      "Processing observatory_id='UMF' - sampling_strategy='soft_sediment' - sheet_type='observatory'\n",
      "Sheet link: https://docs.google.com/spreadsheets/d/1DpQqo2i4q7RAM0sdm_z2pISa8SDBa9aYhWmUflFuUQs/edit?usp=drive_web&ouid=110054986816469121247\n",
      "Sample sheet link: https://docs.google.com/spreadsheets/d/1DpQqo2i4q7RAM0sdm_z2pISa8SDBa9aYhWmUflFuUQs/gviz/tq?tqx=out:csv&sheet=observatory\n",
      "[{'project_name': 'EMOBON', 'latitude': 63.55722, 'longitude': 19.84278, 'geo_loc_name': 'Sweden', 'loc_broad_ocean': 'Baltic Sea', 'loc_broad_ocean_mrgid': 2401, 'loc_regional': 'Gulf of Bothnia', 'loc_regional_mrgid': 2402, 'loc_loc': 'Bothnian Sea', 'loc_loc_mrgid': 19384, 'env_broad_biome': 'marine biome [ENVO:00000447];marine benthic biome[ENVO_01000024]', 'env_local': 'shallow_marine_sediment_[ENVO:03000034]', 'env_package': 'sediment', 'tot_depth_water_col': 13, 'organization': 'Umea Marine Sciences Centre  - Umea University', 'organization_country': 'Sweden', 'organization_edmoid': 622, 'obs_id': 'UMF', 'so_id': 'UMF So', 'extra_site_info': 'This station (NB3) is included in Sweden’s national monitoring program for benthic macrofauna and thus there is an environmental time series since 1995. Also, data from the national pelagic monitoring station is available fom station B3 about 3 nm away.', 'contact_name': 'Annie Marie Cox', 'contact_email': 'annie.cox@umu.se', 'contact_orcid': nan, 'sediment_type': nan, 'ENA_accession_number_umbrella': 'PRJEB51688', 'ENA_accession_number_project': 'PRJEB57914'}]\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/logsheets/UMF_soft_sediment_observatory_validated.csv\n",
      "Observatory LMO lacks valid sheet URL for soft_sediment\n"
     ]
    }
   ],
   "source": [
    "# Validate the Observatory sheets in the EMO-BON-Metadata Google Sheets\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%reload_ext autoreload\n",
    "\n",
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from validation_classes import observatoryModel\n",
    "\n",
    "\n",
    "def get_sheet(sheet_link: str) -> pd.core.frame.DataFrame:\n",
    "    \"\"\"Returns a Pandas dataframe of the 'observatory' sheets\n",
    "    from the observatories' Google Sheets.\n",
    "    \"\"\"\n",
    "    print(f\"Sheet link: {sheet_link}\")\n",
    "    sampling_sheet_base = sheet_link.split(\"/edit\")[0]\n",
    "    sampling_sheet_suffix = \"/gviz/tq?tqx=out:csv&sheet=observatory\"\n",
    "    sample_sheet_link = sampling_sheet_base + sampling_sheet_suffix\n",
    "    print(f\"Sample sheet link: {sample_sheet_link}\")\n",
    "    df = pd.read_csv(sample_sheet_link, encoding=\"utf-8\")\n",
    "    return df\n",
    "\n",
    "\n",
    "def parse_sample_sheets(\n",
    "    sampling_strategy: str, sheet_type: str, addresses: list[str, str]\n",
    ") -> None:\n",
    "    for observatory in addresses:\n",
    "        observatory_id, sheet_link = observatory\n",
    "        # print(f\"Observatory_id {observatory_id} sheet_link {sheet_link}\")\n",
    "        if not isinstance(sheet_link, str):\n",
    "            # print(f\"This is the sheet_link type {type(sheet_link)}\")\n",
    "            if isinstance(sheet_link, float):\n",
    "                if math.isnan(sheet_link):\n",
    "                    print(\n",
    "                        f\"Observatory {observatory_id} lacks \"\n",
    "                        f\"valid sheet URL for {sampling_strategy}\"\n",
    "                    )\n",
    "                    continue\n",
    "            else:\n",
    "                raise ValueError(\n",
    "                    f\"Unknown link {sheet_link} to observatory {observatory_id}\"\n",
    "                )\n",
    "        else:\n",
    "            if observatory_id == \"Plenzia\":\n",
    "                continue  # Sheets not publically available\n",
    "\n",
    "            # if observatory_id not in [\"OSD74\", \"AAOT\"]:\n",
    "            #     continue\n",
    "            # if sampling_strategy != \"water_column\":\n",
    "            #     continue\n",
    "\n",
    "            print(\n",
    "                f\"\\n\\nProcessing {observatory_id=} - {sampling_strategy=} - {sheet_type=}\"\n",
    "            )\n",
    "            df: pd.core.frame.DataFrame = get_sheet(sheet_link)\n",
    "\n",
    "            # Note there is only one row per sheet\n",
    "            data_records_all = df.to_dict(orient=\"records\")\n",
    "\n",
    "            # Get the obs_id from the only row\n",
    "            obs_id = data_records_all[0][\"obs_id\"]\n",
    "            if observatory_id != obs_id:\n",
    "                raise ValueError(f\"Error: {observatory_id=} != {obs_id=}\")\n",
    "\n",
    "            if len(data_records_all) != 1:\n",
    "                raise RuntimeError(f\"Error: {len(data_records_all)} != 1\")\n",
    "\n",
    "            validated_rows = [\n",
    "                observatoryModel(**row).model_dump() for row in data_records_all\n",
    "            ]\n",
    "\n",
    "            save_dir = PROJECT_DIR / \"validated-data\" / \"logsheets\"\n",
    "            outfile_name = (\n",
    "                f\"{observatory_id}_{sampling_strategy}_{sheet_type}_validated.csv\"\n",
    "            )\n",
    "            ndf = pd.DataFrame.from_records(validated_rows, index=\"obs_id\")\n",
    "            out_file = save_dir / outfile_name\n",
    "            ndf.to_csv(out_file)\n",
    "            print(f\"Written {out_file}\")\n",
    "\n",
    "\n",
    "# Get list of observatory ids\n",
    "# df = pd.read_csv(\"./governance/observatories_validated.csv\")\n",
    "# observatory_ids = [id[0] for id in df[[\"observatory_id\"]].values.tolist()]\n",
    "# print(f\"{observatory_ids=}\")\n",
    "\n",
    "# Get list of all URL links to sampling sheets\n",
    "# NB  you cant use a \"with\" closure here when reading the Pandas df\n",
    "df = pd.read_csv(\n",
    "    PROJECT_DIR / \"validated-data\" / \"governance\" / \"logsheets_validated.csv\"\n",
    ")\n",
    "water_column_sheet_addresses = df[[\"observatory_id\", \"water_column\"]].values.tolist()\n",
    "soft_sediment_sheet_addresses = df[[\"observatory_id\", \"soft_sediment\"]].values.tolist()\n",
    "\n",
    "parse_sample_sheets(\"water_column\", \"observatory\", water_column_sheet_addresses)\n",
    "parse_sample_sheets(\"soft_sediment\", \"observatory\", soft_sediment_sheet_addresses)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d3f629d-1b29-46ad-8c81-80084f865c48",
   "metadata": {},
   "source": [
    "# Write the combined Observatory table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0068dff2-16f9-4406-b777-7bf8805894d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "FILES_PATH = PROJECT_DIR / \"validated-data\" / \"logsheets\"\n",
    "\n",
    "\n",
    "def filter_on_sheet_type(csv_file):\n",
    "    return Path(csv_file).stem.split(\"_\")[-2] == \"observatory\"\n",
    "\n",
    "\n",
    "csv_files = [f for f in os.listdir(FILES_PATH) if f.split(\".\")[1] == \"csv\"]\n",
    "observatory_files = list(filter(filter_on_sheet_type, csv_files))\n",
    "\n",
    "frames = []\n",
    "for obs in observatory_files:\n",
    "    df = pd.read_csv(FILES_PATH / obs)\n",
    "    df[\"env_package\"].replace(\"water\", \"water_column\", inplace=True)\n",
    "    df[\"env_package\"].replace(\"sediment\", \"soft_sediment\", inplace=True)\n",
    "    frames.append(df)\n",
    "\n",
    "combined_df = pd.concat(frames)\n",
    "outfile_name = Path(\n",
    "    PROJECT_DIR, \"validated-data\", \"Observatory_combined_logsheets_validated.csv\"\n",
    ")\n",
    "combined_df.to_csv(outfile_name, index=False)\n",
    "combined_df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5c0a641-6fa8-48d5-b3d3-0785beaf2f68",
   "metadata": {},
   "source": [
    "# Write a combined meta-table for each observatory of all validated logsheets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f3cbb42-7bd0-4a45-9ee4-a6d30ce44758",
   "metadata": {},
   "source": [
    "The source_mat_id is the unique key or identifier that links the records in the Batch run_information sheets ([run-information-batch-001.csv](https://raw.githubusercontent.com/emo-bon/sequencing-data/main/shipment/batch-001/run-information-batch-001.csv) and [run-information-batch-002.csv](https://raw.githubusercontent.com/emo-bon/sequencing-data/main/shipment/batch-002/run-information-batch-002.csv)) to the sampling events in the \"sampling\" and \"measured\" sheets of the observatory logsheets (Google Sheets) (e.g. [ESC68N](https://docs.google.com/spreadsheets/d/11_Eu0W1-sDiuzKx1cIl6YuxjRHmWezN6u9v3Ly8JZ3A/edit?gid=0#gid=0)).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "33792f86-8da9-4cd5-9986-29a80aedfbfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Station sampling logsheet: ESC68N_water_column_sampling_validated.csv\n",
      "Observatory ESC68N-water_column has 170 sampling events.\n",
      "Sampling events with no ref_code: 156 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 14 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: Bergen_water_column_sampling_validated.csv\n",
      "Observatory Bergen-water_column has 150 sampling events.\n",
      "Sampling events with no ref_code: 150 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 0 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: MBAL4_water_column_sampling_validated.csv\n",
      "Observatory MBAL4-water_column has 80 sampling events.\n",
      "Sampling events with no ref_code: 72 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 8 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: BPNS_water_column_sampling_validated.csv\n",
      "Observatory BPNS-water_column has 355 sampling events.\n",
      "Sampling events with no ref_code: 337 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 18 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: BPNS_soft_sediment_sampling_validated.csv\n",
      "Observatory BPNS-soft_sediment has 182 sampling events.\n",
      "Sampling events with no ref_code: 174 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 8 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: ROSKOGO_water_column_sampling_validated.csv\n",
      "Observatory ROSKOGO-water_column has 310 sampling events.\n",
      "Sampling events with no ref_code: 292 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 18 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: ROSKOGO_soft_sediment_sampling_validated.csv\n",
      "Observatory ROSKOGO-soft_sediment has 169 sampling events.\n",
      "Sampling events with no ref_code: 163 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 6 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: VB_water_column_sampling_validated.csv\n",
      "Observatory VB-water_column has 824 sampling events.\n",
      "Sampling events with no ref_code: 806 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 18 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: OOB_soft_sediment_sampling_validated.csv\n",
      "Observatory OOB-soft_sediment has 160 sampling events.\n",
      "Sampling events with no ref_code: 153 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 7 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: EMT21_water_column_sampling_validated.csv\n",
      "Observatory EMT21-water_column has 265 sampling events.\n",
      "Sampling events with no ref_code: 251 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 14 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: EMT21_soft_sediment_sampling_validated.csv\n",
      "Observatory EMT21-soft_sediment has 48 sampling events.\n",
      "Sampling events with no ref_code: 48 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 0 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: PiEGetxo_water_column_sampling_validated.csv\n",
      "Observatory PiEGetxo-water_column has 263 sampling events.\n",
      "Sampling events with no ref_code: 249 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 14 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: RFormosa_water_column_sampling_validated.csv\n",
      "Observatory RFormosa-water_column has 180 sampling events.\n",
      "Sampling events with no ref_code: 166 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 14 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: RFormosa_soft_sediment_sampling_validated.csv\n",
      "Observatory RFormosa-soft_sediment has 90 sampling events.\n",
      "Sampling events with no ref_code: 84 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 6 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: OSD74_water_column_sampling_validated.csv\n",
      "Observatory OSD74-water_column has 170 sampling events.\n",
      "Sampling events with no ref_code: 157 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 13 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: AAOT_water_column_sampling_validated.csv\n",
      "Observatory AAOT-water_column has 400 sampling events.\n",
      "Sampling events with no ref_code: 382 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 18 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: NRMCB_water_column_sampling_validated.csv\n",
      "Observatory NRMCB-water_column has 402 sampling events.\n",
      "Sampling events with no ref_code: 384 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 18 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: NRMCB_soft_sediment_sampling_validated.csv\n",
      "Observatory NRMCB-soft_sediment has 100 sampling events.\n",
      "Sampling events with no ref_code: 91 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 9 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: HCMR-1_water_column_sampling_validated.csv\n",
      "Observatory HCMR-1-water_column has 144 sampling events.\n",
      "Sampling events with no ref_code: 134 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 10 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: IUIEilat_water_column_sampling_validated.csv\n",
      "Observatory IUIEilat-water_column has 60 sampling events.\n",
      "Sampling events with no ref_code: 46 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 14 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: UMF_water_column_sampling_validated.csv\n",
      "Observatory UMF-water_column has 154 sampling events.\n",
      "Sampling events with no ref_code: 154 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 0 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: UMF_soft_sediment_sampling_validated.csv\n",
      "Observatory UMF-soft_sediment has 46 sampling events.\n",
      "Sampling events with no ref_code: 46 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 0 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "Station sampling logsheet: LMO_water_column_sampling_validated.csv\n",
      "Observatory LMO-water_column has 120 sampling events.\n",
      "Sampling events with no ref_code: 120 (i.e. they were not sent for sequencing), \n",
      "Sampling events with a ref_code but no measured data with same source_mat_id 0\n",
      "A total of 0 sampling events with refcode and measured sheet were found.\n",
      "\n",
      "There are 227 total ref_codes assigned\n",
      "A total of 0 sequencing event in the Batch 1 & 2 run_information sheets have refcodes that match source_mat_ids in the sample sheets that have duplicate entries\n",
      "Total number of combined sampling events with ref_codes: 227\n",
      "Total number of all_source_mat_ids_from_sheets: 4842 of which 7 were duplicates\n",
      "A total of 0 source_mat_ids in the batch 1 & 2 run information sheets are missing from the observatory sampling sheets\n",
      "\n",
      "Total combined_events: 227 \n",
      "Missing source_mat_ids in combined events: 0 \n",
      "EMO BON ref_codes in run_information sheets: 227 \n",
      "Events ignored due to duplications in sampling sheets: 0 \n",
      "227 + 0 = 227 - 0\n",
      "\n",
      "Sheets are in validated-data/observatories\n"
     ]
    }
   ],
   "source": [
    "import collections\n",
    "import copy\n",
    "import difflib\n",
    "import os\n",
    "import sys\n",
    "\n",
    "import pandas as pd\n",
    "import validators\n",
    "\n",
    "LOGSHEETS_PATH = PROJECT_DIR / \"validated-data\" / \"logsheets\"\n",
    "# These are real duplicates\n",
    "KNOWN_DUPLICATES = []\n",
    "# [\n",
    "#    \"EMOBON_ROSKOGO_Wa_210618_3um_1\",\n",
    "#    \"EMOBON_PiEGetxo_Wa_210824_3um_blank\",\n",
    "# ]\n",
    "\n",
    "\n",
    "def get_observatory_data() -> list[str, str, str]:\n",
    "    # Get list of observatory_ids\n",
    "    df = pd.read_csv(\n",
    "        PROJECT_DIR / \"validated-data\" / \"governance\" / \"logsheets_validated.csv\"\n",
    "    )\n",
    "    observatory_data = df[\n",
    "        [\"observatory_id\", \"water_column\", \"soft_sediment\"]\n",
    "    ].values.tolist()\n",
    "    return observatory_data\n",
    "\n",
    "\n",
    "def get_all_refcodes() -> dict[str, str]:\n",
    "    batch1_run_info_path = \"https://raw.githubusercontent.com/emo-bon/sequencing-data/main/shipment/batch-001/run-information-batch-001.csv\"\n",
    "    batch2_run_info_path = \"https://raw.githubusercontent.com/emo-bon/sequencing-data/main/shipment/batch-002/run-information-batch-002.csv\"\n",
    "    # Get list of batch1 <source_mat_id>, <ref_code>'s\n",
    "    df = pd.read_csv(batch1_run_info_path)\n",
    "    refcodes = {}\n",
    "    for i in df[[\"source_mat_id\", \"ref_code\"]].values.tolist():\n",
    "        if i[0] in refcodes:\n",
    "            raise ValueError(f\"Duplicate source material id {i[0]}\")\n",
    "        refcodes.update(dict([i]))\n",
    "    # Get list of batch2 <source_mat_id>, <ref_code>'s\n",
    "    df = pd.read_csv(batch2_run_info_path)\n",
    "    b2_refcodes = {}\n",
    "    for i in df[[\"source_mat_id\", \"ref_code\"]].values.tolist():\n",
    "        if i[0] in refcodes:\n",
    "            raise ValueError(f\"Duplicate source material id {i[0]}\")\n",
    "        b2_refcodes.update(dict([i]))\n",
    "    refcodes.update(b2_refcodes)\n",
    "    return refcodes\n",
    "\n",
    "\n",
    "def parse_observatory_sample_type(\n",
    "    observatory_id: str,\n",
    "    obs_refcodes: dict[str, str],\n",
    "    sampling_type: str,\n",
    "    save_table: bool = False,\n",
    "    verbose: bool = True,\n",
    ") -> list[dict[str, str]]:\n",
    "    \"\"\"An observatory is an EMBRC station and it has an ID\n",
    "    Each observatory may take either or both of the \"water_column\" and \"soft_sediment\" sampling types\n",
    "    Each sampling type has both a \"sampling\" and \"measured\" sheet\n",
    "\n",
    "    This function returns a list of sampling events each of which is a dict with key/value pairs for each field and value\n",
    "    \"\"\"\n",
    "\n",
    "    sampling_data_filename = f\"{observatory_id}_{sampling_type}_sampling_validated.csv\"\n",
    "    measured_data_filename = f\"{observatory_id}_{sampling_type}_measured_validated.csv\"\n",
    "\n",
    "    sampling_data = pd.read_csv(LOGSHEETS_PATH / sampling_data_filename)\n",
    "    measured_data = pd.read_csv(LOGSHEETS_PATH / measured_data_filename)\n",
    "\n",
    "    sampling_events = sampling_data.to_dict(orient=\"records\")\n",
    "    measured_events = measured_data.to_dict(orient=\"records\")\n",
    "\n",
    "    # To be returned\n",
    "    source_mat_ids_from_combined_events = []  # List of all source_mat_ids\n",
    "    all_sampling_source_mat_ids = sampling_data[\"source_mat_id\"].values.tolist()\n",
    "    missing_measured_but_refcode_present = 0  # Shouldn't happen if sheet is not broken\n",
    "    duplicates_ignored_counter = (\n",
    "        0  # How many known duplicate source_mat_ids in the sheets are we ignoring\n",
    "    )\n",
    "    combined_events = []  # List of all sampling/measured events\n",
    "\n",
    "    # Internal\n",
    "    no_refcode_counter = (\n",
    "        0  # Sampling events without a refcode ie not sent to sequencing\n",
    "    )\n",
    "    source_mat_ids_in_sampling_with_refcode_missing_from_measured = []  # Again shouldn't happen\n",
    "    refcodes_in_run_info = []  # List of refcodes matched to source_mat_ids\n",
    "\n",
    "    for sampling_event in sampling_events:\n",
    "        # Checking consistency\n",
    "        # Does this sampling event have a ref_code\n",
    "        # If yes, then it should have both a sampling and measured sheet\n",
    "        # If no, we can ignore it\n",
    "        event_mat_id = sampling_event[\"source_mat_id\"]\n",
    "\n",
    "        # Two source_mat_id's in the Batch 1 & 2 run_informations match duplicate sampling events\n",
    "        # in the sampling sheets - here we ignore those two:\n",
    "        if event_mat_id in KNOWN_DUPLICATES:\n",
    "            duplicates_ignored_counter += 1\n",
    "            # print(f\"IGNORING DUP: {event_mat_id}\")\n",
    "            continue\n",
    "\n",
    "        # Hack for HCMR-1 replicates which are \"blank1\" and \"blank2\" in the sampling sheet\n",
    "        # become just blank in the measured and run_information sheet from where we get the refcodes\n",
    "        # if event_mat_id == \"EMOBON_HCMR-1_Wa_210917_3um_blank1\":\n",
    "        #    event_mat_id = \"EMOBON_HCMR-1_Wa_210917_3um_blank\"\n",
    "        # if event_mat_id == \"EMOBON_HCMR-1_Wa_210917_0.2um_blank1\":\n",
    "        #    event_mat_id = \"EMOBON_HCMR-1_Wa_210917_0.2um_blank\"\n",
    "\n",
    "        try:\n",
    "            refcode = obs_refcodes[event_mat_id]\n",
    "        except KeyError:\n",
    "            no_refcode_counter += 1\n",
    "            # OK so has not been sent to sequencing; ignore\n",
    "            continue\n",
    "\n",
    "        event_measured = False\n",
    "        for measured_event in measured_events:\n",
    "            try:\n",
    "                measured_event[\"source_mat_id\"]\n",
    "            except KeyError as e:\n",
    "                raise KeyError(f\"Key error: {measured_event}\") from e\n",
    "                # Should not happen\n",
    "            if measured_event[\"source_mat_id\"] == event_mat_id:\n",
    "                event_measured = copy.deepcopy(measured_event)\n",
    "                break\n",
    "\n",
    "        if not event_measured:\n",
    "            # sampling sheet source_mat_id has ref_code in run_information\n",
    "            # but the corresponding measured sheet lacks the same sources_mat_id\n",
    "            # This shouldn't happen unless the auto-formatting of the 'source_mat_id'\n",
    "            # field in the 'measured' sheet is broken - which is exactly what happened.\n",
    "            missing_measured_but_refcode_present += 1\n",
    "            source_mat_ids_in_sampling_with_refcode_missing_from_measured.append(\n",
    "                event_mat_id\n",
    "            )\n",
    "            continue\n",
    "        else:\n",
    "            sampling_event[\"ref_code\"] = refcode  # key to sequence data\n",
    "            sampling_event[\"obs_id\"] = observatory_id  # key to observatory data\n",
    "            if refcode in refcodes_in_run_info:\n",
    "                raise ValueError(\n",
    "                    f\"Error: {refcode=} match more that one sampling event \"\n",
    "                    f\"with the {source_mat_id=}\"\n",
    "                )\n",
    "            else:\n",
    "                refcodes_in_run_info.append(refcode)\n",
    "                # Delete the now duplicated source_mat_id in the combined event\n",
    "                del event_measured[\"source_mat_id\"]\n",
    "                source_mat_ids_from_combined_events.append(event_mat_id)\n",
    "                sampling_event.update(event_measured)\n",
    "                combined_events.append(sampling_event)\n",
    "\n",
    "    if verbose:\n",
    "        print(\n",
    "            f\"Station sampling logsheet: {sampling_data_filename}\\n\"\n",
    "            f\"Observatory {observatory_id}-{sampling_type} has {len(sampling_events)} sampling events.\\n\"\n",
    "            f\"Sampling events with no ref_code: {no_refcode_counter} (i.e. they were not sent for sequencing), \\n\"\n",
    "            f\"Sampling events with a ref_code but no measured data with same source_mat_id {missing_measured_but_refcode_present}\\n\"\n",
    "            f\"A total of {len(combined_events)} sampling events with refcode and measured sheet were found.\\n\"\n",
    "        )\n",
    "\n",
    "    # Did we find all the sampling events?\n",
    "    se = len(sampling_events)\n",
    "    ce = len(combined_events)\n",
    "    mmbrcp = missing_measured_but_refcode_present\n",
    "    nrc = no_refcode_counter\n",
    "    dc = duplicates_ignored_counter\n",
    "    if se != (ce + nrc + mmbrcp + dc):\n",
    "        print(\n",
    "            f\"Something is a foot: len(sampling_events) {se} != \"\n",
    "            f\"(len(combined_events) {ce}\"\n",
    "            f\"+ no_refcode_counter {nrc}\"\n",
    "            f\"+ missing_measured_but_refcode_present {mmbrcp})\"\n",
    "            f\"+ known duplicates {dc} \\n\"\n",
    "            \"CAUTION: could be due to missing source_mat_ids in the sampling sheets \\n\"\n",
    "            \" -- check output below \\n\"\n",
    "        )\n",
    "\n",
    "    if len(combined_events) != 0 and save_table:\n",
    "        save_dir = PROJECT_DIR / \"validated-data\" / \"observatories\"\n",
    "        outfile_name = f\"{observatory_id}_{sampling_type}_combined_validated.csv\"\n",
    "        ndf = pd.DataFrame.from_records(combined_events, index=\"source_mat_id\")\n",
    "        ndf.to_csv(os.path.join(save_dir, outfile_name))\n",
    "\n",
    "    return (\n",
    "        source_mat_ids_from_combined_events,\n",
    "        all_sampling_source_mat_ids,\n",
    "        missing_measured_but_refcode_present,\n",
    "        duplicates_ignored_counter,\n",
    "        combined_events,\n",
    "    )\n",
    "\n",
    "\n",
    "def validate_observatories(\n",
    "    observatory_data, obs_refcodes, save_table=False\n",
    ") -> tuple[list, list, int, int, list]:\n",
    "    # To be returned\n",
    "    all_source_mat_ids_from_combined_events: list[str] = []\n",
    "    all_source_mat_ids_from_sheets: list[str] = []\n",
    "    all_missing_measured_but_refcode_present: int = 0\n",
    "    all_duplicates_ignored: int = 0\n",
    "    all_combined_events: list[list[str, ...]] = []\n",
    "\n",
    "    for observatory_id, water_column, soft_sediment in observatory_data:\n",
    "        # Plenzia still broken\n",
    "        if observatory_id == \"Plenzia\":\n",
    "            continue\n",
    "\n",
    "        observatories_present = []\n",
    "        if validators.url(water_column):\n",
    "            observatories_present.append(\"water_column\")\n",
    "        if validators.url(soft_sediment):\n",
    "            observatories_present.append(\"soft_sediment\")\n",
    "\n",
    "        for sampling_strategy in observatories_present:\n",
    "            r = parse_observatory_sample_type(\n",
    "                observatory_id, obs_refcodes, sampling_strategy, save_table\n",
    "            )\n",
    "\n",
    "            (all_source_mat_ids_from_combined_events.extend(r[0]),)\n",
    "            (all_source_mat_ids_from_sheets.extend(r[1]),)\n",
    "            # Cannot use += on tuple unpackinghttp://ceta.ualg.pt:8000/\n",
    "            all_missing_measured_but_refcode_present = (\n",
    "                all_missing_measured_but_refcode_present + r[2]\n",
    "            )\n",
    "            all_duplicates_ignored = all_duplicates_ignored + r[3]\n",
    "            all_combined_events.extend(r[4])\n",
    "\n",
    "    return (\n",
    "        all_source_mat_ids_from_combined_events,\n",
    "        all_source_mat_ids_from_sheets,\n",
    "        all_missing_measured_but_refcode_present,\n",
    "        all_duplicates_ignored,\n",
    "        all_combined_events,\n",
    "    )\n",
    "\n",
    "\n",
    "############ VALIDATE OBSERVATORIES ###################################################\n",
    "\n",
    "# Get observatory data and ref_codes\n",
    "observatory_data: list[str, str, str] = get_observatory_data()\n",
    "obs_refcodes: dict[str, str] = get_all_refcodes()\n",
    "\n",
    "result = validate_observatories(observatory_data, obs_refcodes, save_table=True)\n",
    "\n",
    "all_source_mat_ids_from_combined_events = result[0]\n",
    "all_source_mat_ids_from_sheets = result[1]\n",
    "all_missing_measured_but_refcode_present = result[2]\n",
    "all_duplicates_ignored = result[3]\n",
    "all_combined_events = result[4]\n",
    "\n",
    "############## REPORT STATISTICS ########################################################\n",
    "\n",
    "############### REFCODES ####################################\n",
    "print(f\"There are {len(obs_refcodes)} total ref_codes assigned\")\n",
    "# We are ignoring both of the duplicates but it's only 1 record expected missing from total_combined events\n",
    "\n",
    "############## DUPLICATES ###################################\n",
    "total_dups_ignored = int(all_duplicates_ignored / 2)\n",
    "print(\n",
    "    f\"A total of {total_dups_ignored} sequencing event in the Batch 1 & 2 run_information sheets \"\n",
    "    f\"have refcodes that match source_mat_ids in the sample sheets that have duplicate entries\"\n",
    ")\n",
    "print(\n",
    "    f\"Total number of combined sampling events with ref_codes: {len(all_combined_events)}\"\n",
    ")\n",
    "duplicates = [\n",
    "    source_mat_id\n",
    "    for source_mat_id, count in collections.Counter(\n",
    "        all_source_mat_ids_from_sheets\n",
    "    ).items()\n",
    "    if count > 1\n",
    "]\n",
    "print(\n",
    "    f\"Total number of all_source_mat_ids_from_sheets: {len(all_source_mat_ids_from_sheets)}\"\n",
    "    f\" of which {len(duplicates)} were duplicates\"\n",
    ")\n",
    "\n",
    "############# SOURCE_MAT_IDS ################################\n",
    "missing_source_mat_ids = []\n",
    "for source_mat_id in obs_refcodes:\n",
    "    # source_mat_ids are the keys in the refcode dict of the run_information\n",
    "    # print(f\"refcode from run_information: {refcode}\")\n",
    "\n",
    "    # Hack for HCMR-1\n",
    "    # if source_mat_id == \"EMOBON_HCMR-1_Wa_210917_3um_blank\":\n",
    "    #    source_mat_id = \"EMOBON_HCMR-1_Wa_210917_3um_blank1\"\n",
    "    # if source_mat_id == \"EMOBON_HCMR-1_Wa_210917_0.2um_blank\":\n",
    "    #    source_mat_id = \"EMOBON_HCMR-1_Wa_210917_0.2um_blank1\"\n",
    "\n",
    "    if source_mat_id not in all_source_mat_ids_from_sheets:\n",
    "        # print(f\"source_mat_id {source_mat_id} is missing from the sampling sheets\")\n",
    "\n",
    "        # Get close matches to missing source_mat_id\n",
    "        matches = difflib.get_close_matches(\n",
    "            source_mat_id, all_source_mat_ids_from_sheets, n=3\n",
    "        )\n",
    "        missing_source_mat_ids.append([source_mat_id, matches])\n",
    "\n",
    "if missing_source_mat_ids:\n",
    "    print(\"\\n\\nThe missing source_mat_ids that are in the run information sheets are:\")\n",
    "    for missing in missing_source_mat_ids:\n",
    "        join = \" \".join(missing[1])\n",
    "        print(f\"Missing source_mat_id is {missing[0]} close matches are \\n\\t {join}\")\n",
    "\n",
    "    for missing in missing_source_mat_ids:\n",
    "        print(missing)\n",
    "\n",
    "print(\n",
    "    f\"A total of {len(missing_source_mat_ids)} source_mat_ids \"\n",
    "    f\"in the batch 1 & 2 run information sheets are missing from the \"\n",
    "    f\"observatory sampling sheets\"\n",
    ")\n",
    "\n",
    "missing = False  # CAUTION: THIS SHOULD BE ZERO!\n",
    "counter = 0\n",
    "# TODO: take all_source_mat_ids_from_combined_events directly from all_combined_events rather than having\n",
    "# a separate list\n",
    "for source_mat_id in all_source_mat_ids_from_combined_events:\n",
    "    # source_mat_ids are the keys in the refcode dict of the run_information\n",
    "    # print(f\"source_mat_id: {source_mat_id}\")\n",
    "    if source_mat_id not in obs_refcodes:\n",
    "        # print(f\"source_mat_id {source_mat_id} is missing from the run_information\")\n",
    "        counter += 1\n",
    "        missing = True\n",
    "if missing:\n",
    "    print(\n",
    "        f\"ERROR: A total of {counter} source_mat_ids \"\n",
    "        f\"in the sampling sheets that also have refcodes are missing from the \"\n",
    "        f\"Batch 1 & 2 run information sheets\"\n",
    "    )\n",
    "\n",
    "##################### SUMMARY #################################\n",
    "total = len(all_combined_events) + len(missing_source_mat_ids)\n",
    "out_dir = Path(\"validated-data\", \"observatories\")\n",
    "print(\n",
    "    f\"\\nTotal combined_events: {len(all_combined_events)} \\n\"\n",
    "    f\"Missing source_mat_ids in combined events: {len(missing_source_mat_ids)} \\n\"\n",
    "    f\"EMO BON ref_codes in run_information sheets: {len(obs_refcodes)} \\n\"\n",
    "    f\"Events ignored due to duplications in sampling sheets: {total_dups_ignored} \\n\"\n",
    "    f\"{len(all_combined_events)} + {len(missing_source_mat_ids)} = {len(obs_refcodes)} - {total_dups_ignored}\\n\"\n",
    "    f\"\\nSheets are in {out_dir}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29b5e06b-a1b8-4661-afd2-3ae2a42b3103",
   "metadata": {},
   "source": [
    "# Create the single metadata table from the combined observatories tables"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1b0522c-c143-42f4-8d0f-db82b0e56e3f",
   "metadata": {},
   "source": [
    "Combined observatories metadata tables are in validated_data/observatories and have the file name format:\n",
    "`<observatory_id>_<sampling_strategy>_combined_validated.csv`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc9344de",
   "metadata": {},
   "source": [
    "This needs to be more sophisticated; stations have removed many fields so you cant just combine them as the fields dont match up between sheets:\n",
    "\n",
    "- Get a list of all fields from the Definitions sub-sheet of the sampling sheets and measured sheets\n",
    "- Loop over fields when adding them to the combined sheets inserting Null values in fields that are missing\n",
    "- Identify any fields that are not in the definitions (should not be any...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1c07f7b5-9c18-4e11-9ee0-2e8061603b6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number_of_rows=227\n",
      "Written /home/cymon/vscode/git-repos/emo-bon-data-validation/validated-data/Batch1and2_combined_logsheets_2024-11-12.csv\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 227 entries, 0 to 13\n",
      "Columns: 114 entries, source_mat_id to env_package\n",
      "dtypes: Int64(1), float64(61), object(52)\n",
      "memory usage: 204.2+ KB\n"
     ]
    }
   ],
   "source": [
    "FILE_PATH = PROJECT_DIR / \"validated-data\" / \"observatories\"\n",
    "csv_files = [f for f in os.listdir(FILE_PATH) if f.split(\".\")[1] == \"csv\"]\n",
    "\n",
    "number_of_rows = 0\n",
    "frames = []\n",
    "for obs in csv_files:\n",
    "    front = obs.split(\"_combined_validated.csv\")[0]\n",
    "    obs_id, strategy = front.split(\"_\", 1)\n",
    "    # print(f\"{obs_id=} -> - {strategy=}\")\n",
    "    df = pd.read_csv(\n",
    "        os.path.join(FILE_PATH, obs), dtype={\"tax_id\": \"Int64\"}\n",
    "    )  # See note above\n",
    "    for index, _ in df.iterrows():\n",
    "        # This is the name given in the Observatory sheet Google Spreadsheet\n",
    "        df.loc[index, \"env_package\"] = strategy\n",
    "        number_of_rows += 1\n",
    "    frames.append(df)\n",
    "\n",
    "print(f\"{number_of_rows=}\")\n",
    "combined_df = pd.concat(frames)\n",
    "today = datetime.datetime.now().strftime(\"%Y-%m-%d\")\n",
    "outfile_name = f\"Batch1and2_combined_logsheets_{today}.csv\"\n",
    "out_path = PROJECT_DIR / \"validated-data\" / outfile_name\n",
    "combined_df.to_csv(out_path, index=False)\n",
    "print(f\"Written {out_path}\")\n",
    "combined_df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a41b2851-bde9-49b7-9051-8cd847368008",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
